{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "a8ca2e5c",
   "metadata": {},
   "source": [
    "## CCLE\n",
    "\n",
    "## Description\n",
    "\n",
    "This CCLE dataset is made by performing the usual bulk RNA-seq technique, then normalizing using RPKM (Reads Per Kilobase of transcript per Million) to reduce variance and make samples comparable.\n",
    "\n",
    "The RPKM normalization solves two problems created during sequencing:\n",
    "\n",
    "1. **Sequencing depth**, which happens when samples are sequenced to different depths. Some samples may have more or fewer total reads, but that difference does not reflect biology, so we correct for it.  \n",
    "2. **Gene length**, because longer genes generate more reads just by being longer, and shorter genes fewer reads. Normalizing by gene length accounts for that.\n",
    "\n",
    "The normalized \\((i, j)\\) RPKM value, where \\(i\\) is a gene and \\(j\\) is a sample, is:\n",
    "\n",
    "$$\n",
    "\\mathrm{RPKM}_{i,j}\n",
    "=\n",
    "\\frac{x_{i,j}}{\\,l_i \\cdot \\sum_k x_{k,j}\\,}\n",
    "\\times 10^6\n",
    "$$\n",
    "\n",
    "<!-- blank line above is important -->\n",
    "- \\(x_{i,j}\\) is the raw read count for gene \\(i\\) in sample \\(j\\)  \n",
    "- \\(l_i\\) is the length of gene \\(i\\) in kilobases (kb)  \n",
    "- \\(\\sum_{k} x_{k,j}\\) is the total reads in sample \\(j\\)  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "3403de5a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "67e0b620",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ---------mRNA---------\n",
    "# Description:\n",
    "# Expression Data with gene as index and cells/samples as columns\n",
    "data_mrna_seq_rpkm = pd.read_csv('ccle_data/data_mrna_seq_rpkm.txt',\n",
    "                    sep = '\\t',\n",
    "                    comment = '#')\n",
    "\n",
    "data_mrna_seq_rpkm.set_index('Hugo_Symbol',inplace=True)\n",
    "\n",
    "# Merge with mean duplicated rows\n",
    "data_mrna_seq_rpkm = data_mrna_seq_rpkm.groupby(data_mrna_seq_rpkm.index).mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "a8916bbb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>22RV1_PROSTATE</th>\n",
       "      <th>2313287_STOMACH</th>\n",
       "      <th>253JBV_URINARY_TRACT</th>\n",
       "      <th>253J_URINARY_TRACT</th>\n",
       "      <th>42MGBA_CENTRAL_NERVOUS_SYSTEM</th>\n",
       "      <th>5637_URINARY_TRACT</th>\n",
       "      <th>59M_OVARY</th>\n",
       "      <th>639V_URINARY_TRACT</th>\n",
       "      <th>647V_URINARY_TRACT</th>\n",
       "      <th>697_HAEMATOPOIETIC_AND_LYMPHOID_TISSUE</th>\n",
       "      <th>...</th>\n",
       "      <th>UMUC16_URINARY_TRACT</th>\n",
       "      <th>UMUC4_URINARY_TRACT</th>\n",
       "      <th>UMUC5_URINARY_TRACT</th>\n",
       "      <th>UMUC6_URINARY_TRACT</th>\n",
       "      <th>UMUC7_URINARY_TRACT</th>\n",
       "      <th>UMUC9_URINARY_TRACT</th>\n",
       "      <th>UPCISCC152_UPPER_AERODIGESTIVE_TRACT</th>\n",
       "      <th>UW228_CENTRAL_NERVOUS_SYSTEM</th>\n",
       "      <th>Y79_AUTONOMIC_GANGLIA</th>\n",
       "      <th>YAMATO_SOFT_TISSUE</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CRIPTOP1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7SK</th>\n",
       "      <td>0.029785</td>\n",
       "      <td>0.03377</td>\n",
       "      <td>0.047608</td>\n",
       "      <td>0.02094</td>\n",
       "      <td>0.00430</td>\n",
       "      <td>0.017847</td>\n",
       "      <td>0.006893</td>\n",
       "      <td>0.00704</td>\n",
       "      <td>0.008042</td>\n",
       "      <td>0.148955</td>\n",
       "      <td>...</td>\n",
       "      <td>0.083175</td>\n",
       "      <td>0.02683</td>\n",
       "      <td>0.002032</td>\n",
       "      <td>0.024993</td>\n",
       "      <td>0.094365</td>\n",
       "      <td>0.393628</td>\n",
       "      <td>0.035012</td>\n",
       "      <td>0.002465</td>\n",
       "      <td>0.05875</td>\n",
       "      <td>0.067953</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>0.362300</td>\n",
       "      <td>0.00608</td>\n",
       "      <td>0.115170</td>\n",
       "      <td>0.33737</td>\n",
       "      <td>3.06452</td>\n",
       "      <td>0.013390</td>\n",
       "      <td>2.180160</td>\n",
       "      <td>2.24186</td>\n",
       "      <td>0.095400</td>\n",
       "      <td>3.698050</td>\n",
       "      <td>...</td>\n",
       "      <td>0.012700</td>\n",
       "      <td>0.09324</td>\n",
       "      <td>0.011340</td>\n",
       "      <td>0.060090</td>\n",
       "      <td>0.026380</td>\n",
       "      <td>0.019230</td>\n",
       "      <td>0.505580</td>\n",
       "      <td>0.578600</td>\n",
       "      <td>0.71115</td>\n",
       "      <td>0.849120</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1BG-AS1</th>\n",
       "      <td>4.815880</td>\n",
       "      <td>0.18959</td>\n",
       "      <td>0.751510</td>\n",
       "      <td>1.32578</td>\n",
       "      <td>9.15532</td>\n",
       "      <td>0.064590</td>\n",
       "      <td>1.619220</td>\n",
       "      <td>4.48351</td>\n",
       "      <td>0.631410</td>\n",
       "      <td>6.483210</td>\n",
       "      <td>...</td>\n",
       "      <td>0.371360</td>\n",
       "      <td>0.52954</td>\n",
       "      <td>0.187460</td>\n",
       "      <td>0.121010</td>\n",
       "      <td>0.359240</td>\n",
       "      <td>0.809130</td>\n",
       "      <td>2.750650</td>\n",
       "      <td>2.918060</td>\n",
       "      <td>2.06849</td>\n",
       "      <td>2.596400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1CF</th>\n",
       "      <td>5.654860</td>\n",
       "      <td>2.55482</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.12239</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.011540</td>\n",
       "      <td>0.00589</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 1156 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             22RV1_PROSTATE  2313287_STOMACH  253JBV_URINARY_TRACT  \\\n",
       "Hugo_Symbol                                                          \n",
       " CRIPTOP1          0.000000          0.00000              0.000000   \n",
       "7SK                0.029785          0.03377              0.047608   \n",
       "A1BG               0.362300          0.00608              0.115170   \n",
       "A1BG-AS1           4.815880          0.18959              0.751510   \n",
       "A1CF               5.654860          2.55482              0.000000   \n",
       "\n",
       "             253J_URINARY_TRACT  42MGBA_CENTRAL_NERVOUS_SYSTEM  \\\n",
       "Hugo_Symbol                                                      \n",
       " CRIPTOP1               0.00000                        0.00000   \n",
       "7SK                     0.02094                        0.00430   \n",
       "A1BG                    0.33737                        3.06452   \n",
       "A1BG-AS1                1.32578                        9.15532   \n",
       "A1CF                    0.00000                        0.00000   \n",
       "\n",
       "             5637_URINARY_TRACT  59M_OVARY  639V_URINARY_TRACT  \\\n",
       "Hugo_Symbol                                                      \n",
       " CRIPTOP1              0.000000   0.000000             0.00000   \n",
       "7SK                    0.017847   0.006893             0.00704   \n",
       "A1BG                   0.013390   2.180160             2.24186   \n",
       "A1BG-AS1               0.064590   1.619220             4.48351   \n",
       "A1CF                   0.000000   0.000000             0.00000   \n",
       "\n",
       "             647V_URINARY_TRACT  697_HAEMATOPOIETIC_AND_LYMPHOID_TISSUE  ...  \\\n",
       "Hugo_Symbol                                                              ...   \n",
       " CRIPTOP1              0.000000                                0.000000  ...   \n",
       "7SK                    0.008042                                0.148955  ...   \n",
       "A1BG                   0.095400                                3.698050  ...   \n",
       "A1BG-AS1               0.631410                                6.483210  ...   \n",
       "A1CF                   0.000000                                0.000000  ...   \n",
       "\n",
       "             UMUC16_URINARY_TRACT  UMUC4_URINARY_TRACT  UMUC5_URINARY_TRACT  \\\n",
       "Hugo_Symbol                                                                   \n",
       " CRIPTOP1                0.000000              0.00000             0.000000   \n",
       "7SK                      0.083175              0.02683             0.002032   \n",
       "A1BG                     0.012700              0.09324             0.011340   \n",
       "A1BG-AS1                 0.371360              0.52954             0.187460   \n",
       "A1CF                     0.000000              4.12239             0.000000   \n",
       "\n",
       "             UMUC6_URINARY_TRACT  UMUC7_URINARY_TRACT  UMUC9_URINARY_TRACT  \\\n",
       "Hugo_Symbol                                                                  \n",
       " CRIPTOP1               0.000000             0.000000             0.000000   \n",
       "7SK                     0.024993             0.094365             0.393628   \n",
       "A1BG                    0.060090             0.026380             0.019230   \n",
       "A1BG-AS1                0.121010             0.359240             0.809130   \n",
       "A1CF                    0.000000             0.000000             0.000000   \n",
       "\n",
       "             UPCISCC152_UPPER_AERODIGESTIVE_TRACT  \\\n",
       "Hugo_Symbol                                         \n",
       " CRIPTOP1                                0.000000   \n",
       "7SK                                      0.035012   \n",
       "A1BG                                     0.505580   \n",
       "A1BG-AS1                                 2.750650   \n",
       "A1CF                                     0.000000   \n",
       "\n",
       "             UW228_CENTRAL_NERVOUS_SYSTEM  Y79_AUTONOMIC_GANGLIA  \\\n",
       "Hugo_Symbol                                                        \n",
       " CRIPTOP1                        0.000000                0.00000   \n",
       "7SK                              0.002465                0.05875   \n",
       "A1BG                             0.578600                0.71115   \n",
       "A1BG-AS1                         2.918060                2.06849   \n",
       "A1CF                             0.011540                0.00589   \n",
       "\n",
       "             YAMATO_SOFT_TISSUE  \n",
       "Hugo_Symbol                      \n",
       " CRIPTOP1              0.000000  \n",
       "7SK                    0.067953  \n",
       "A1BG                   0.849120  \n",
       "A1BG-AS1               2.596400  \n",
       "A1CF                   0.000000  \n",
       "\n",
       "[5 rows x 1156 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_mrna_seq_rpkm.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "8301ef8b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#--------Mut query----------\n",
    "# Description:\n",
    "# Data telling  which sample is mutated and not-mutated\n",
    "mutations= pd.read_csv('ccle_data/mutations.txt',\n",
    "                    sep = '\\t',\n",
    "                    comment = '#')\n",
    "mutations.set_index('SAMPLE_ID',inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "072c2041",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>STUDY_ID</th>\n",
       "      <th>TP53</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMPLE_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SJRH30_SOFT_TISSUE</th>\n",
       "      <td>ccle_broad_2019</td>\n",
       "      <td>R280S R273C Y205C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FADU_UPPER_AERODIGESTIVE_TRACT</th>\n",
       "      <td>ccle_broad_2019</td>\n",
       "      <td>R248L X225_splice</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HH_HAEMATOPOIETIC_AND_LYMPHOID_TISSUE</th>\n",
       "      <td>ccle_broad_2019</td>\n",
       "      <td>X187_splice</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SKNMC_BONE</th>\n",
       "      <td>ccle_broad_2019</td>\n",
       "      <td>WT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SNU182_LIVER</th>\n",
       "      <td>ccle_broad_2019</td>\n",
       "      <td>S215I</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              STUDY_ID               TP53\n",
       "SAMPLE_ID                                                                \n",
       "SJRH30_SOFT_TISSUE                     ccle_broad_2019  R280S R273C Y205C\n",
       "FADU_UPPER_AERODIGESTIVE_TRACT         ccle_broad_2019  R248L X225_splice\n",
       "HH_HAEMATOPOIETIC_AND_LYMPHOID_TISSUE  ccle_broad_2019        X187_splice\n",
       "SKNMC_BONE                             ccle_broad_2019                 WT\n",
       "SNU182_LIVER                           ccle_broad_2019              S215I"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mutations.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "006ae621",
   "metadata": {},
   "source": [
    "## Why predict on just the five Variant_Type classes\n",
    "\n",
    "We focus on SNP, DNP, ONP, INS, and DEL because:\n",
    "\n",
    "1. Biological clarity: these five labels describe the fundamental mutation mechanism, single-base or multi-base changes, insertions, and deletions, so the model learns clear patterns.  \n",
    "2. Balanced data: each Variant_Type occurs often enough to give the model enough examples, while the detailed Variant_Classification labels are very uneven and would leave some classes too small to learn.  \n",
    "3. Reduced complexity: Variant_Classification depends on gene structure and reading frame (for example a SNP can be silent or missense depending on codon), which our sequence‐only model cannot infer without extra annotation.  \n",
    "4. Modular workflow: once the model tags a variant as INS or DEL, we can apply separate rules or a second model to predict functional impact, keeping each step simpler and more reliable.  \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "e215f51c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/xz/cl0qfkz56dl2q1b8vhqj2v180000gn/T/ipykernel_14233/449069273.py:3: DtypeWarning: Columns (4,51,53,54,56,57) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  data_mutations = pd.read_csv('ccle_data/data_mutations.txt',\n"
     ]
    }
   ],
   "source": [
    "# ------Mut ALL-----------\n",
    "# Dataset containing mutation classes for all genes, it only contains mutated samples\n",
    "data_mutations = pd.read_csv('ccle_data/data_mutations.txt',\n",
    "                    sep = '\\t',\n",
    "                    comment = '#')\n",
    "\n",
    "# Extract TP53 from all genes\n",
    "data_mutations = data_mutations[data_mutations['Hugo_Symbol'] == 'TP53']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "c240f890",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['SNP' 'DNP' 'DEL' 'INS']\n"
     ]
    }
   ],
   "source": [
    "unique_labels = data_mutations['Variant_Type'].unique()\n",
    "print(unique_labels)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "7f29c365",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Missense_Mutation' 'Nonsense_Mutation' 'Splice_Site' 'Silent'\n",
      " 'Splice_Region' 'Frame_Shift_Del' 'Frame_Shift_Ins' 'In_Frame_Del'\n",
      " 'In_Frame_Ins']\n"
     ]
    }
   ],
   "source": [
    "unique_labels = data_mutations['Variant_Classification'].unique()\n",
    "print(unique_labels)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "34cf6889",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remove unwanted information\n",
    "data_mutations = data_mutations[['Tumor_Sample_Barcode', 'Variant_Type']]\n",
    "data_mutations.set_index('Tumor_Sample_Barcode', inplace=True)\n",
    "# There are repetitions of my mutation type (Variant_Type)\n",
    "# if there is the same sample with different Variant_Type it should be removed\n",
    "variant_check = data_mutations.groupby(data_mutations.index)[\"Variant_Type\"].nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "748e4345",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Variant_Type</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tumor_Sample_Barcode</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>22RV1_PROSTATE</th>\n",
       "      <td>SNP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A431_SKIN</th>\n",
       "      <td>SNP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A4FUK_HAEMATOPOIETIC_AND_LYMPHOID_TISSUE</th>\n",
       "      <td>SNP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BICR16_UPPER_AERODIGESTIVE_TRACT</th>\n",
       "      <td>SNP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BICR78_UPPER_AERODIGESTIVE_TRACT</th>\n",
       "      <td>SNP</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         Variant_Type\n",
       "Tumor_Sample_Barcode                                 \n",
       "22RV1_PROSTATE                                    SNP\n",
       "A431_SKIN                                         SNP\n",
       "A4FUK_HAEMATOPOIETIC_AND_LYMPHOID_TISSUE          SNP\n",
       "BICR16_UPPER_AERODIGESTIVE_TRACT                  SNP\n",
       "BICR78_UPPER_AERODIGESTIVE_TRACT                  SNP"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_mutations.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "714a1aa5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SNP: 1042\n",
      "DEL: 116\n",
      "INS: 34\n",
      "DNP: 26\n"
     ]
    }
   ],
   "source": [
    "# Count how many mutations are in each Variant_Type\n",
    "counts = data_mutations['Variant_Type'].value_counts()\n",
    "\n",
    "# Print them one per line\n",
    "for variant_type, n in counts.items():\n",
    "    print(f\"{variant_type}: {n}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "965665f0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>CRIPTOP1</th>\n",
       "      <th>7SK</th>\n",
       "      <th>A1BG</th>\n",
       "      <th>A1BG-AS1</th>\n",
       "      <th>A1CF</th>\n",
       "      <th>A2M</th>\n",
       "      <th>A2M-AS1</th>\n",
       "      <th>A2ML1</th>\n",
       "      <th>A2ML1-AS1</th>\n",
       "      <th>A2ML1-AS2</th>\n",
       "      <th>...</th>\n",
       "      <th>snoZ13_snr52</th>\n",
       "      <th>snoZ178</th>\n",
       "      <th>snoZ185</th>\n",
       "      <th>snoZ247</th>\n",
       "      <th>snoZ40</th>\n",
       "      <th>snoZ5</th>\n",
       "      <th>snoZ6</th>\n",
       "      <th>snosnR60_Z15</th>\n",
       "      <th>snosnR66</th>\n",
       "      <th>yR211F11.2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>22RV1_PROSTATE</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.029785</td>\n",
       "      <td>0.36230</td>\n",
       "      <td>4.81588</td>\n",
       "      <td>5.65486</td>\n",
       "      <td>1.98954</td>\n",
       "      <td>1.27348</td>\n",
       "      <td>0.01960</td>\n",
       "      <td>0.91070</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.341940</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0755</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2313287_STOMACH</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.033770</td>\n",
       "      <td>0.00608</td>\n",
       "      <td>0.18959</td>\n",
       "      <td>2.55482</td>\n",
       "      <td>0.00782</td>\n",
       "      <td>0.22274</td>\n",
       "      <td>0.01738</td>\n",
       "      <td>0.32801</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.239685</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>253JBV_URINARY_TRACT</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.047608</td>\n",
       "      <td>0.11517</td>\n",
       "      <td>0.75151</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.05044</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00544</td>\n",
       "      <td>0.60473</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>253J_URINARY_TRACT</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.020940</td>\n",
       "      <td>0.33737</td>\n",
       "      <td>1.32578</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.02878</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00547</td>\n",
       "      <td>0.46686</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42MGBA_CENTRAL_NERVOUS_SYSTEM</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.004300</td>\n",
       "      <td>3.06452</td>\n",
       "      <td>9.15532</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.02477</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.09900</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.000913</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 54353 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Hugo_Symbol                     CRIPTOP1       7SK     A1BG  A1BG-AS1  \\\n",
       "22RV1_PROSTATE                       0.0  0.029785  0.36230   4.81588   \n",
       "2313287_STOMACH                      0.0  0.033770  0.00608   0.18959   \n",
       "253JBV_URINARY_TRACT                 0.0  0.047608  0.11517   0.75151   \n",
       "253J_URINARY_TRACT                   0.0  0.020940  0.33737   1.32578   \n",
       "42MGBA_CENTRAL_NERVOUS_SYSTEM        0.0  0.004300  3.06452   9.15532   \n",
       "\n",
       "Hugo_Symbol                       A1CF      A2M  A2M-AS1    A2ML1  A2ML1-AS1  \\\n",
       "22RV1_PROSTATE                 5.65486  1.98954  1.27348  0.01960    0.91070   \n",
       "2313287_STOMACH                2.55482  0.00782  0.22274  0.01738    0.32801   \n",
       "253JBV_URINARY_TRACT           0.00000  0.05044  0.00000  0.00544    0.60473   \n",
       "253J_URINARY_TRACT             0.00000  0.02878  0.00000  0.00547    0.46686   \n",
       "42MGBA_CENTRAL_NERVOUS_SYSTEM  0.00000  0.02477  0.00000  0.00000    0.09900   \n",
       "\n",
       "Hugo_Symbol                    A2ML1-AS2  ...  snoZ13_snr52  snoZ178  snoZ185  \\\n",
       "22RV1_PROSTATE                       0.0  ...           0.0      0.0      0.0   \n",
       "2313287_STOMACH                      0.0  ...           0.0      0.0      0.0   \n",
       "253JBV_URINARY_TRACT                 0.0  ...           0.0      0.0      0.0   \n",
       "253J_URINARY_TRACT                   0.0  ...           0.0      0.0      0.0   \n",
       "42MGBA_CENTRAL_NERVOUS_SYSTEM        0.0  ...           0.0      0.0      0.0   \n",
       "\n",
       "Hugo_Symbol                    snoZ247  snoZ40  snoZ5     snoZ6  snosnR60_Z15  \\\n",
       "22RV1_PROSTATE                     0.0     0.0    0.0  3.341940           0.0   \n",
       "2313287_STOMACH                    0.0     0.0    0.0  3.239685           0.0   \n",
       "253JBV_URINARY_TRACT               0.0     0.0    0.0  0.000000           0.0   \n",
       "253J_URINARY_TRACT                 0.0     0.0    0.0  0.000000           0.0   \n",
       "42MGBA_CENTRAL_NERVOUS_SYSTEM      0.0     0.0    0.0  1.000913           0.0   \n",
       "\n",
       "Hugo_Symbol                    snosnR66  yR211F11.2  \n",
       "22RV1_PROSTATE                      0.0      0.0755  \n",
       "2313287_STOMACH                     0.0      0.0000  \n",
       "253JBV_URINARY_TRACT                0.0      0.0000  \n",
       "253J_URINARY_TRACT                  0.0      0.0000  \n",
       "42MGBA_CENTRAL_NERVOUS_SYSTEM       0.0      0.0000  \n",
       "\n",
       "[5 rows x 54353 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_t = data_mrna_seq_rpkm.T\n",
    "\n",
    "data_t.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "9fe7496e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Hugo_Symbol\n",
       " CRIPTOP1       0.000000\n",
       "7SK             0.047608\n",
       "A1BG            0.115170\n",
       "A1BG-AS1        0.751510\n",
       "A1CF            0.000000\n",
       "                  ...   \n",
       "snoZ5           0.000000\n",
       "snoZ6           0.000000\n",
       "snosnR60_Z15    0.000000\n",
       "snosnR66        0.000000\n",
       "yR211F11.2      0.000000\n",
       "Name: 253JBV_URINARY_TRACT, Length: 54353, dtype: float64"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_t.loc['253JBV_URINARY_TRACT']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "cad8e5cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "code = {\"SNP\": 0, \"DNP\": 1, \"DEL\": 2, \"INS\": 3}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "a01e0551",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of samples discarded: 287\n"
     ]
    }
   ],
   "source": [
    "# Build target vector y\n",
    "y = []\n",
    "\n",
    "# prepare lists to collect rows and their sample name\n",
    "X_rows = []\n",
    "sample_names = []\n",
    "c = 0\n",
    "\n",
    "# iterate over each mutation record\n",
    "for bc, mut in data_mutations.iterrows():\n",
    "    # check if this barcode is in data_t’s index\n",
    "    if bc in data_t.index:\n",
    "        # grab the full row from data_t and store it\n",
    "        X_rows.append(data_t.loc[bc].values)\n",
    "        y.append(code[mut['Variant_Type']])\n",
    "        sample_names.append(bc)\n",
    "    else:\n",
    "        c += 1\n",
    "\n",
    "print(f\"Number of samples discarded: {c}\")\n",
    "\n",
    "# build a new DataFrame X from the collected rows\n",
    "X = pd.DataFrame(\n",
    "    X_rows,\n",
    "    index=sample_names,\n",
    "    columns=data_t.columns\n",
    ")\n",
    "\n",
    "y = np.array(y)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "d9c47f9c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "X shape: (931, 54353)\n",
      "y shape: (931,)\n"
     ]
    }
   ],
   "source": [
    "print(f\"X shape: {X.shape}\")\n",
    "print(f\"y shape: {y.shape}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "97f30bde",
   "metadata": {},
   "source": [
    "**Step 2: Train - Test split (80% - 20%)**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "da468c7d",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.model_selection import train_test_split"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "76b0a91f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Train shape:\n",
      "\tX_train: (744, 54353)\n",
      "\ty_train: (744,)\n",
      "Test shape:\n",
      "\tX_test: (187, 54353)\n",
      "\ty_test: (187,)\n"
     ]
    }
   ],
   "source": [
    "# 80% train, 20% test, stratify to preserve class balance\n",
    "X_train, X_test, y_train, y_test = train_test_split(\n",
    "    X, y,\n",
    "    test_size=0.2,\n",
    "    random_state=42,\n",
    "    stratify=y\n",
    ")\n",
    "\n",
    "print(f\"Train shape:\\n\\tX_train: {X_train.shape}\\n\\ty_train: {y_train.shape}\")\n",
    "print(f\"Test shape:\\n\\tX_test: {X_test.shape}\\n\\ty_test: {y_test.shape}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "48156d5d",
   "metadata": {},
   "source": [
    "**Step 3: Model selection and Training**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "99a1fd12",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.ensemble import RandomForestClassifier\n",
    "from sklearn.metrics import classification_report, confusion_matrix, roc_auc_score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "9ce43c68",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Instantiate the model\n",
    "rf = RandomForestClassifier(\n",
    "    n_estimators=500,      # you can tune this\n",
    "    max_depth=None,        # full depth; you can limit for speed/regularization\n",
    "    random_state=42,\n",
    "    n_jobs=-1              # use all cores\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "04bb89dc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>#sk-container-id-1 {\n",
       "  /* Definition of color scheme common for light and dark mode */\n",
       "  --sklearn-color-text: #000;\n",
       "  --sklearn-color-text-muted: #666;\n",
       "  --sklearn-color-line: gray;\n",
       "  /* Definition of color scheme for unfitted estimators */\n",
       "  --sklearn-color-unfitted-level-0: #fff5e6;\n",
       "  --sklearn-color-unfitted-level-1: #f6e4d2;\n",
       "  --sklearn-color-unfitted-level-2: #ffe0b3;\n",
       "  --sklearn-color-unfitted-level-3: chocolate;\n",
       "  /* Definition of color scheme for fitted estimators */\n",
       "  --sklearn-color-fitted-level-0: #f0f8ff;\n",
       "  --sklearn-color-fitted-level-1: #d4ebff;\n",
       "  --sklearn-color-fitted-level-2: #b3dbfd;\n",
       "  --sklearn-color-fitted-level-3: cornflowerblue;\n",
       "\n",
       "  /* Specific color for light theme */\n",
       "  --sklearn-color-text-on-default-background: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, black)));\n",
       "  --sklearn-color-background: var(--sg-background-color, var(--theme-background, var(--jp-layout-color0, white)));\n",
       "  --sklearn-color-border-box: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, black)));\n",
       "  --sklearn-color-icon: #696969;\n",
       "\n",
       "  @media (prefers-color-scheme: dark) {\n",
       "    /* Redefinition of color scheme for dark theme */\n",
       "    --sklearn-color-text-on-default-background: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, white)));\n",
       "    --sklearn-color-background: var(--sg-background-color, var(--theme-background, var(--jp-layout-color0, #111)));\n",
       "    --sklearn-color-border-box: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, white)));\n",
       "    --sklearn-color-icon: #878787;\n",
       "  }\n",
       "}\n",
       "\n",
       "#sk-container-id-1 {\n",
       "  color: var(--sklearn-color-text);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 pre {\n",
       "  padding: 0;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 input.sk-hidden--visually {\n",
       "  border: 0;\n",
       "  clip: rect(1px 1px 1px 1px);\n",
       "  clip: rect(1px, 1px, 1px, 1px);\n",
       "  height: 1px;\n",
       "  margin: -1px;\n",
       "  overflow: hidden;\n",
       "  padding: 0;\n",
       "  position: absolute;\n",
       "  width: 1px;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-dashed-wrapped {\n",
       "  border: 1px dashed var(--sklearn-color-line);\n",
       "  margin: 0 0.4em 0.5em 0.4em;\n",
       "  box-sizing: border-box;\n",
       "  padding-bottom: 0.4em;\n",
       "  background-color: var(--sklearn-color-background);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-container {\n",
       "  /* jupyter's `normalize.less` sets `[hidden] { display: none; }`\n",
       "     but bootstrap.min.css set `[hidden] { display: none !important; }`\n",
       "     so we also need the `!important` here to be able to override the\n",
       "     default hidden behavior on the sphinx rendered scikit-learn.org.\n",
       "     See: https://github.com/scikit-learn/scikit-learn/issues/21755 */\n",
       "  display: inline-block !important;\n",
       "  position: relative;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-text-repr-fallback {\n",
       "  display: none;\n",
       "}\n",
       "\n",
       "div.sk-parallel-item,\n",
       "div.sk-serial,\n",
       "div.sk-item {\n",
       "  /* draw centered vertical line to link estimators */\n",
       "  background-image: linear-gradient(var(--sklearn-color-text-on-default-background), var(--sklearn-color-text-on-default-background));\n",
       "  background-size: 2px 100%;\n",
       "  background-repeat: no-repeat;\n",
       "  background-position: center center;\n",
       "}\n",
       "\n",
       "/* Parallel-specific style estimator block */\n",
       "\n",
       "#sk-container-id-1 div.sk-parallel-item::after {\n",
       "  content: \"\";\n",
       "  width: 100%;\n",
       "  border-bottom: 2px solid var(--sklearn-color-text-on-default-background);\n",
       "  flex-grow: 1;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-parallel {\n",
       "  display: flex;\n",
       "  align-items: stretch;\n",
       "  justify-content: center;\n",
       "  background-color: var(--sklearn-color-background);\n",
       "  position: relative;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-parallel-item {\n",
       "  display: flex;\n",
       "  flex-direction: column;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-parallel-item:first-child::after {\n",
       "  align-self: flex-end;\n",
       "  width: 50%;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-parallel-item:last-child::after {\n",
       "  align-self: flex-start;\n",
       "  width: 50%;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-parallel-item:only-child::after {\n",
       "  width: 0;\n",
       "}\n",
       "\n",
       "/* Serial-specific style estimator block */\n",
       "\n",
       "#sk-container-id-1 div.sk-serial {\n",
       "  display: flex;\n",
       "  flex-direction: column;\n",
       "  align-items: center;\n",
       "  background-color: var(--sklearn-color-background);\n",
       "  padding-right: 1em;\n",
       "  padding-left: 1em;\n",
       "}\n",
       "\n",
       "\n",
       "/* Toggleable style: style used for estimator/Pipeline/ColumnTransformer box that is\n",
       "clickable and can be expanded/collapsed.\n",
       "- Pipeline and ColumnTransformer use this feature and define the default style\n",
       "- Estimators will overwrite some part of the style using the `sk-estimator` class\n",
       "*/\n",
       "\n",
       "/* Pipeline and ColumnTransformer style (default) */\n",
       "\n",
       "#sk-container-id-1 div.sk-toggleable {\n",
       "  /* Default theme specific background. It is overwritten whether we have a\n",
       "  specific estimator or a Pipeline/ColumnTransformer */\n",
       "  background-color: var(--sklearn-color-background);\n",
       "}\n",
       "\n",
       "/* Toggleable label */\n",
       "#sk-container-id-1 label.sk-toggleable__label {\n",
       "  cursor: pointer;\n",
       "  display: flex;\n",
       "  width: 100%;\n",
       "  margin-bottom: 0;\n",
       "  padding: 0.5em;\n",
       "  box-sizing: border-box;\n",
       "  text-align: center;\n",
       "  align-items: start;\n",
       "  justify-content: space-between;\n",
       "  gap: 0.5em;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 label.sk-toggleable__label .caption {\n",
       "  font-size: 0.6rem;\n",
       "  font-weight: lighter;\n",
       "  color: var(--sklearn-color-text-muted);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 label.sk-toggleable__label-arrow:before {\n",
       "  /* Arrow on the left of the label */\n",
       "  content: \"▸\";\n",
       "  float: left;\n",
       "  margin-right: 0.25em;\n",
       "  color: var(--sklearn-color-icon);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 label.sk-toggleable__label-arrow:hover:before {\n",
       "  color: var(--sklearn-color-text);\n",
       "}\n",
       "\n",
       "/* Toggleable content - dropdown */\n",
       "\n",
       "#sk-container-id-1 div.sk-toggleable__content {\n",
       "  max-height: 0;\n",
       "  max-width: 0;\n",
       "  overflow: hidden;\n",
       "  text-align: left;\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-0);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-toggleable__content.fitted {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-0);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-toggleable__content pre {\n",
       "  margin: 0.2em;\n",
       "  border-radius: 0.25em;\n",
       "  color: var(--sklearn-color-text);\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-0);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-toggleable__content.fitted pre {\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-0);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 input.sk-toggleable__control:checked~div.sk-toggleable__content {\n",
       "  /* Expand drop-down */\n",
       "  max-height: 200px;\n",
       "  max-width: 100%;\n",
       "  overflow: auto;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 input.sk-toggleable__control:checked~label.sk-toggleable__label-arrow:before {\n",
       "  content: \"▾\";\n",
       "}\n",
       "\n",
       "/* Pipeline/ColumnTransformer-specific style */\n",
       "\n",
       "#sk-container-id-1 div.sk-label input.sk-toggleable__control:checked~label.sk-toggleable__label {\n",
       "  color: var(--sklearn-color-text);\n",
       "  background-color: var(--sklearn-color-unfitted-level-2);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-label.fitted input.sk-toggleable__control:checked~label.sk-toggleable__label {\n",
       "  background-color: var(--sklearn-color-fitted-level-2);\n",
       "}\n",
       "\n",
       "/* Estimator-specific style */\n",
       "\n",
       "/* Colorize estimator box */\n",
       "#sk-container-id-1 div.sk-estimator input.sk-toggleable__control:checked~label.sk-toggleable__label {\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-2);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-estimator.fitted input.sk-toggleable__control:checked~label.sk-toggleable__label {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-2);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-label label.sk-toggleable__label,\n",
       "#sk-container-id-1 div.sk-label label {\n",
       "  /* The background is the default theme color */\n",
       "  color: var(--sklearn-color-text-on-default-background);\n",
       "}\n",
       "\n",
       "/* On hover, darken the color of the background */\n",
       "#sk-container-id-1 div.sk-label:hover label.sk-toggleable__label {\n",
       "  color: var(--sklearn-color-text);\n",
       "  background-color: var(--sklearn-color-unfitted-level-2);\n",
       "}\n",
       "\n",
       "/* Label box, darken color on hover, fitted */\n",
       "#sk-container-id-1 div.sk-label.fitted:hover label.sk-toggleable__label.fitted {\n",
       "  color: var(--sklearn-color-text);\n",
       "  background-color: var(--sklearn-color-fitted-level-2);\n",
       "}\n",
       "\n",
       "/* Estimator label */\n",
       "\n",
       "#sk-container-id-1 div.sk-label label {\n",
       "  font-family: monospace;\n",
       "  font-weight: bold;\n",
       "  display: inline-block;\n",
       "  line-height: 1.2em;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-label-container {\n",
       "  text-align: center;\n",
       "}\n",
       "\n",
       "/* Estimator-specific */\n",
       "#sk-container-id-1 div.sk-estimator {\n",
       "  font-family: monospace;\n",
       "  border: 1px dotted var(--sklearn-color-border-box);\n",
       "  border-radius: 0.25em;\n",
       "  box-sizing: border-box;\n",
       "  margin-bottom: 0.5em;\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-0);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-estimator.fitted {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-0);\n",
       "}\n",
       "\n",
       "/* on hover */\n",
       "#sk-container-id-1 div.sk-estimator:hover {\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-2);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-estimator.fitted:hover {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-2);\n",
       "}\n",
       "\n",
       "/* Specification for estimator info (e.g. \"i\" and \"?\") */\n",
       "\n",
       "/* Common style for \"i\" and \"?\" */\n",
       "\n",
       ".sk-estimator-doc-link,\n",
       "a:link.sk-estimator-doc-link,\n",
       "a:visited.sk-estimator-doc-link {\n",
       "  float: right;\n",
       "  font-size: smaller;\n",
       "  line-height: 1em;\n",
       "  font-family: monospace;\n",
       "  background-color: var(--sklearn-color-background);\n",
       "  border-radius: 1em;\n",
       "  height: 1em;\n",
       "  width: 1em;\n",
       "  text-decoration: none !important;\n",
       "  margin-left: 0.5em;\n",
       "  text-align: center;\n",
       "  /* unfitted */\n",
       "  border: var(--sklearn-color-unfitted-level-1) 1pt solid;\n",
       "  color: var(--sklearn-color-unfitted-level-1);\n",
       "}\n",
       "\n",
       ".sk-estimator-doc-link.fitted,\n",
       "a:link.sk-estimator-doc-link.fitted,\n",
       "a:visited.sk-estimator-doc-link.fitted {\n",
       "  /* fitted */\n",
       "  border: var(--sklearn-color-fitted-level-1) 1pt solid;\n",
       "  color: var(--sklearn-color-fitted-level-1);\n",
       "}\n",
       "\n",
       "/* On hover */\n",
       "div.sk-estimator:hover .sk-estimator-doc-link:hover,\n",
       ".sk-estimator-doc-link:hover,\n",
       "div.sk-label-container:hover .sk-estimator-doc-link:hover,\n",
       ".sk-estimator-doc-link:hover {\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-3);\n",
       "  color: var(--sklearn-color-background);\n",
       "  text-decoration: none;\n",
       "}\n",
       "\n",
       "div.sk-estimator.fitted:hover .sk-estimator-doc-link.fitted:hover,\n",
       ".sk-estimator-doc-link.fitted:hover,\n",
       "div.sk-label-container:hover .sk-estimator-doc-link.fitted:hover,\n",
       ".sk-estimator-doc-link.fitted:hover {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-3);\n",
       "  color: var(--sklearn-color-background);\n",
       "  text-decoration: none;\n",
       "}\n",
       "\n",
       "/* Span, style for the box shown on hovering the info icon */\n",
       ".sk-estimator-doc-link span {\n",
       "  display: none;\n",
       "  z-index: 9999;\n",
       "  position: relative;\n",
       "  font-weight: normal;\n",
       "  right: .2ex;\n",
       "  padding: .5ex;\n",
       "  margin: .5ex;\n",
       "  width: min-content;\n",
       "  min-width: 20ex;\n",
       "  max-width: 50ex;\n",
       "  color: var(--sklearn-color-text);\n",
       "  box-shadow: 2pt 2pt 4pt #999;\n",
       "  /* unfitted */\n",
       "  background: var(--sklearn-color-unfitted-level-0);\n",
       "  border: .5pt solid var(--sklearn-color-unfitted-level-3);\n",
       "}\n",
       "\n",
       ".sk-estimator-doc-link.fitted span {\n",
       "  /* fitted */\n",
       "  background: var(--sklearn-color-fitted-level-0);\n",
       "  border: var(--sklearn-color-fitted-level-3);\n",
       "}\n",
       "\n",
       ".sk-estimator-doc-link:hover span {\n",
       "  display: block;\n",
       "}\n",
       "\n",
       "/* \"?\"-specific style due to the `<a>` HTML tag */\n",
       "\n",
       "#sk-container-id-1 a.estimator_doc_link {\n",
       "  float: right;\n",
       "  font-size: 1rem;\n",
       "  line-height: 1em;\n",
       "  font-family: monospace;\n",
       "  background-color: var(--sklearn-color-background);\n",
       "  border-radius: 1rem;\n",
       "  height: 1rem;\n",
       "  width: 1rem;\n",
       "  text-decoration: none;\n",
       "  /* unfitted */\n",
       "  color: var(--sklearn-color-unfitted-level-1);\n",
       "  border: var(--sklearn-color-unfitted-level-1) 1pt solid;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 a.estimator_doc_link.fitted {\n",
       "  /* fitted */\n",
       "  border: var(--sklearn-color-fitted-level-1) 1pt solid;\n",
       "  color: var(--sklearn-color-fitted-level-1);\n",
       "}\n",
       "\n",
       "/* On hover */\n",
       "#sk-container-id-1 a.estimator_doc_link:hover {\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-3);\n",
       "  color: var(--sklearn-color-background);\n",
       "  text-decoration: none;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 a.estimator_doc_link.fitted:hover {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-3);\n",
       "}\n",
       "</style><div id=\"sk-container-id-1\" class=\"sk-top-container\"><div class=\"sk-text-repr-fallback\"><pre>RandomForestClassifier(n_estimators=500, n_jobs=-1, random_state=42)</pre><b>In a Jupyter environment, please rerun this cell to show the HTML representation or trust the notebook. <br />On GitHub, the HTML representation is unable to render, please try loading this page with nbviewer.org.</b></div><div class=\"sk-container\" hidden><div class=\"sk-item\"><div class=\"sk-estimator fitted sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-1\" type=\"checkbox\" checked><label for=\"sk-estimator-id-1\" class=\"sk-toggleable__label fitted sk-toggleable__label-arrow\"><div><div>RandomForestClassifier</div></div><div><a class=\"sk-estimator-doc-link fitted\" rel=\"noreferrer\" target=\"_blank\" href=\"https://scikit-learn.org/1.6/modules/generated/sklearn.ensemble.RandomForestClassifier.html\">?<span>Documentation for RandomForestClassifier</span></a><span class=\"sk-estimator-doc-link fitted\">i<span>Fitted</span></span></div></label><div class=\"sk-toggleable__content fitted\"><pre>RandomForestClassifier(n_estimators=500, n_jobs=-1, random_state=42)</pre></div> </div></div></div></div>"
      ],
      "text/plain": [
       "RandomForestClassifier(n_estimators=500, n_jobs=-1, random_state=42)"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Train\n",
    "rf.fit(X_train, y_train)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "4c8246cf",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/lorenzoturiano/Bocconi/ML-lab/ml_env/lib/python3.10/site-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but RandomForestClassifier was fitted with feature names\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "ename": "RuntimeError",
     "evalue": "Numpy is not available",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mRuntimeError\u001b[0m                              Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[33], line 2\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;66;03m# Predict\u001b[39;00m\n\u001b[0;32m----> 2\u001b[0m y_pred \u001b[38;5;241m=\u001b[39m \u001b[43mrf\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mpredict\u001b[49m\u001b[43m(\u001b[49m\u001b[43mX_test\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m      4\u001b[0m n \u001b[38;5;241m=\u001b[39m y_test\u001b[38;5;241m.\u001b[39mshape[\u001b[38;5;241m0\u001b[39m]\n\u001b[1;32m      5\u001b[0m count_mispred \u001b[38;5;241m=\u001b[39m \u001b[38;5;241m0\u001b[39m\n",
      "File \u001b[0;32m~/Bocconi/ML-lab/ml_env/lib/python3.10/site-packages/sklearn/ensemble/_forest.py:904\u001b[0m, in \u001b[0;36mForestClassifier.predict\u001b[0;34m(self, X)\u001b[0m\n\u001b[1;32m    883\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21mpredict\u001b[39m(\u001b[38;5;28mself\u001b[39m, X):\n\u001b[1;32m    884\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m    885\u001b[0m \u001b[38;5;124;03m    Predict class for X.\u001b[39;00m\n\u001b[1;32m    886\u001b[0m \n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    902\u001b[0m \u001b[38;5;124;03m        The predicted classes.\u001b[39;00m\n\u001b[1;32m    903\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[0;32m--> 904\u001b[0m     proba \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mpredict_proba\u001b[49m\u001b[43m(\u001b[49m\u001b[43mX\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    906\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mn_outputs_ \u001b[38;5;241m==\u001b[39m \u001b[38;5;241m1\u001b[39m:\n\u001b[1;32m    907\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mclasses_\u001b[38;5;241m.\u001b[39mtake(np\u001b[38;5;241m.\u001b[39margmax(proba, axis\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m1\u001b[39m), axis\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m0\u001b[39m)\n",
      "File \u001b[0;32m~/Bocconi/ML-lab/ml_env/lib/python3.10/site-packages/sklearn/ensemble/_forest.py:946\u001b[0m, in \u001b[0;36mForestClassifier.predict_proba\u001b[0;34m(self, X)\u001b[0m\n\u001b[1;32m    944\u001b[0m check_is_fitted(\u001b[38;5;28mself\u001b[39m)\n\u001b[1;32m    945\u001b[0m \u001b[38;5;66;03m# Check data\u001b[39;00m\n\u001b[0;32m--> 946\u001b[0m X \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_validate_X_predict\u001b[49m\u001b[43m(\u001b[49m\u001b[43mX\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    948\u001b[0m \u001b[38;5;66;03m# Assign chunk of trees to jobs\u001b[39;00m\n\u001b[1;32m    949\u001b[0m n_jobs, _, _ \u001b[38;5;241m=\u001b[39m _partition_estimators(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mn_estimators, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mn_jobs)\n",
      "File \u001b[0;32m~/Bocconi/ML-lab/ml_env/lib/python3.10/site-packages/sklearn/ensemble/_forest.py:638\u001b[0m, in \u001b[0;36mBaseForest._validate_X_predict\u001b[0;34m(self, X)\u001b[0m\n\u001b[1;32m    635\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    636\u001b[0m     ensure_all_finite \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m\n\u001b[0;32m--> 638\u001b[0m X \u001b[38;5;241m=\u001b[39m \u001b[43mvalidate_data\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    639\u001b[0m \u001b[43m    \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    640\u001b[0m \u001b[43m    \u001b[49m\u001b[43mX\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    641\u001b[0m \u001b[43m    \u001b[49m\u001b[43mdtype\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mDTYPE\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    642\u001b[0m \u001b[43m    \u001b[49m\u001b[43maccept_sparse\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mcsr\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    643\u001b[0m \u001b[43m    \u001b[49m\u001b[43mreset\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m    644\u001b[0m \u001b[43m    \u001b[49m\u001b[43mensure_all_finite\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mensure_all_finite\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    645\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    646\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m issparse(X) \u001b[38;5;129;01mand\u001b[39;00m (X\u001b[38;5;241m.\u001b[39mindices\u001b[38;5;241m.\u001b[39mdtype \u001b[38;5;241m!=\u001b[39m np\u001b[38;5;241m.\u001b[39mintc \u001b[38;5;129;01mor\u001b[39;00m X\u001b[38;5;241m.\u001b[39mindptr\u001b[38;5;241m.\u001b[39mdtype \u001b[38;5;241m!=\u001b[39m np\u001b[38;5;241m.\u001b[39mintc):\n\u001b[1;32m    647\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mNo support for np.int64 index based sparse matrices\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "File \u001b[0;32m~/Bocconi/ML-lab/ml_env/lib/python3.10/site-packages/sklearn/utils/validation.py:2944\u001b[0m, in \u001b[0;36mvalidate_data\u001b[0;34m(_estimator, X, y, reset, validate_separately, skip_check_array, **check_params)\u001b[0m\n\u001b[1;32m   2942\u001b[0m         out \u001b[38;5;241m=\u001b[39m X, y\n\u001b[1;32m   2943\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m no_val_X \u001b[38;5;129;01mand\u001b[39;00m no_val_y:\n\u001b[0;32m-> 2944\u001b[0m     out \u001b[38;5;241m=\u001b[39m \u001b[43mcheck_array\u001b[49m\u001b[43m(\u001b[49m\u001b[43mX\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43minput_name\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mX\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mcheck_params\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   2945\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m no_val_X \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m no_val_y:\n\u001b[1;32m   2946\u001b[0m     out \u001b[38;5;241m=\u001b[39m _check_y(y, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mcheck_params)\n",
      "File \u001b[0;32m~/Bocconi/ML-lab/ml_env/lib/python3.10/site-packages/sklearn/utils/validation.py:1055\u001b[0m, in \u001b[0;36mcheck_array\u001b[0;34m(array, accept_sparse, accept_large_sparse, dtype, order, copy, force_writeable, force_all_finite, ensure_all_finite, ensure_non_negative, ensure_2d, allow_nd, ensure_min_samples, ensure_min_features, estimator, input_name)\u001b[0m\n\u001b[1;32m   1053\u001b[0m         array \u001b[38;5;241m=\u001b[39m xp\u001b[38;5;241m.\u001b[39mastype(array, dtype, copy\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m)\n\u001b[1;32m   1054\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m-> 1055\u001b[0m         array \u001b[38;5;241m=\u001b[39m \u001b[43m_asarray_with_order\u001b[49m\u001b[43m(\u001b[49m\u001b[43marray\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43morder\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43morder\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdtype\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdtype\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mxp\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mxp\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1056\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m ComplexWarning \u001b[38;5;28;01mas\u001b[39;00m complex_warning:\n\u001b[1;32m   1057\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\n\u001b[1;32m   1058\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mComplex data not supported\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;132;01m{}\u001b[39;00m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;241m.\u001b[39mformat(array)\n\u001b[1;32m   1059\u001b[0m     ) \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01mcomplex_warning\u001b[39;00m\n",
      "File \u001b[0;32m~/Bocconi/ML-lab/ml_env/lib/python3.10/site-packages/sklearn/utils/_array_api.py:839\u001b[0m, in \u001b[0;36m_asarray_with_order\u001b[0;34m(array, dtype, order, copy, xp, device)\u001b[0m\n\u001b[1;32m    837\u001b[0m     array \u001b[38;5;241m=\u001b[39m numpy\u001b[38;5;241m.\u001b[39marray(array, order\u001b[38;5;241m=\u001b[39morder, dtype\u001b[38;5;241m=\u001b[39mdtype)\n\u001b[1;32m    838\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m--> 839\u001b[0m     array \u001b[38;5;241m=\u001b[39m \u001b[43mnumpy\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43masarray\u001b[49m\u001b[43m(\u001b[49m\u001b[43marray\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43morder\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43morder\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdtype\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdtype\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    841\u001b[0m \u001b[38;5;66;03m# At this point array is a NumPy ndarray. We convert it to an array\u001b[39;00m\n\u001b[1;32m    842\u001b[0m \u001b[38;5;66;03m# container that is consistent with the input's namespace.\u001b[39;00m\n\u001b[1;32m    843\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m xp\u001b[38;5;241m.\u001b[39masarray(array)\n",
      "File \u001b[0;32m~/Bocconi/ML-lab/ml_env/lib/python3.10/site-packages/torch/_tensor.py:1064\u001b[0m, in \u001b[0;36mTensor.__array__\u001b[0;34m(self, dtype)\u001b[0m\n\u001b[1;32m   1062\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mnumpy()\n\u001b[1;32m   1063\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m-> 1064\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mnumpy\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241m.\u001b[39mastype(dtype, copy\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m)\n",
      "\u001b[0;31mRuntimeError\u001b[0m: Numpy is not available"
     ]
    }
   ],
   "source": [
    "# Predict\n",
    "y_pred = rf.predict(X_test)\n",
    "\n",
    "n = y_test.shape[0]\n",
    "count_mispred = 0\n",
    "for i in range(n):\n",
    "    if y_test[i] != y_pred[i]:\n",
    "        count_mispred += 1\n",
    "\n",
    "# Compute the percentage of mispredictions (accuracy)\n",
    "percentage_mispred = count_mispred / n\n",
    "print(f'Accuracy: {(1-percentage_mispred)*100:.2f}%')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "feb7612d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "A module that was compiled using NumPy 1.x cannot be run in\n",
      "NumPy 2.2.6 as it may crash. To support both 1.x and 2.x\n",
      "versions of NumPy, modules must be compiled with NumPy 2.0.\n",
      "Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.\n",
      "\n",
      "If you are a user of the module, the easiest solution will be to\n",
      "downgrade to 'numpy<2' or try to upgrade the affected module.\n",
      "We expect that some modules will need time to support NumPy 2.\n",
      "\n",
      "Traceback (most recent call last):  File \"/usr/local/Cellar/python@3.10/3.10.17/Frameworks/Python.framework/Versions/3.10/lib/python3.10/runpy.py\", line 196, in _run_module_as_main\n",
      "    return _run_code(code, main_globals, None,\n",
      "  File \"/usr/local/Cellar/python@3.10/3.10.17/Frameworks/Python.framework/Versions/3.10/lib/python3.10/runpy.py\", line 86, in _run_code\n",
      "    exec(code, run_globals)\n",
      "  File \"/Users/lorenzoturiano/Bocconi/ML-lab/ml_env/lib/python3.10/site-packages/ipykernel_launcher.py\", line 18, in <module>\n",
      "    app.launch_new_instance()\n",
      "  File \"/Users/lorenzoturiano/Bocconi/ML-lab/ml_env/lib/python3.10/site-packages/traitlets/config/application.py\", line 1075, in launch_instance\n",
      "    app.start()\n",
      "  File \"/Users/lorenzoturiano/Bocconi/ML-lab/ml_env/lib/python3.10/site-packages/ipykernel/kernelapp.py\", line 739, in start\n",
      "    self.io_loop.start()\n",
      "  File \"/Users/lorenzoturiano/Bocconi/ML-lab/ml_env/lib/python3.10/site-packages/tornado/platform/asyncio.py\", line 211, in start\n",
      "    self.asyncio_loop.run_forever()\n",
      "  File \"/usr/local/Cellar/python@3.10/3.10.17/Frameworks/Python.framework/Versions/3.10/lib/python3.10/asyncio/base_events.py\", line 603, in run_forever\n",
      "    self._run_once()\n",
      "  File \"/usr/local/Cellar/python@3.10/3.10.17/Frameworks/Python.framework/Versions/3.10/lib/python3.10/asyncio/base_events.py\", line 1909, in _run_once\n",
      "    handle._run()\n",
      "  File \"/usr/local/Cellar/python@3.10/3.10.17/Frameworks/Python.framework/Versions/3.10/lib/python3.10/asyncio/events.py\", line 80, in _run\n",
      "    self._context.run(self._callback, *self._args)\n",
      "  File \"/Users/lorenzoturiano/Bocconi/ML-lab/ml_env/lib/python3.10/site-packages/ipykernel/kernelbase.py\", line 545, in dispatch_queue\n",
      "    await self.process_one()\n",
      "  File \"/Users/lorenzoturiano/Bocconi/ML-lab/ml_env/lib/python3.10/site-packages/ipykernel/kernelbase.py\", line 534, in process_one\n",
      "    await dispatch(*args)\n",
      "  File \"/Users/lorenzoturiano/Bocconi/ML-lab/ml_env/lib/python3.10/site-packages/ipykernel/kernelbase.py\", line 437, in dispatch_shell\n",
      "    await result\n",
      "  File \"/Users/lorenzoturiano/Bocconi/ML-lab/ml_env/lib/python3.10/site-packages/ipykernel/ipkernel.py\", line 362, in execute_request\n",
      "    await super().execute_request(stream, ident, parent)\n",
      "  File \"/Users/lorenzoturiano/Bocconi/ML-lab/ml_env/lib/python3.10/site-packages/ipykernel/kernelbase.py\", line 778, in execute_request\n",
      "    reply_content = await reply_content\n",
      "  File \"/Users/lorenzoturiano/Bocconi/ML-lab/ml_env/lib/python3.10/site-packages/ipykernel/ipkernel.py\", line 449, in do_execute\n",
      "    res = shell.run_cell(\n",
      "  File \"/Users/lorenzoturiano/Bocconi/ML-lab/ml_env/lib/python3.10/site-packages/ipykernel/zmqshell.py\", line 549, in run_cell\n",
      "    return super().run_cell(*args, **kwargs)\n",
      "  File \"/Users/lorenzoturiano/Bocconi/ML-lab/ml_env/lib/python3.10/site-packages/IPython/core/interactiveshell.py\", line 3077, in run_cell\n",
      "    result = self._run_cell(\n",
      "  File \"/Users/lorenzoturiano/Bocconi/ML-lab/ml_env/lib/python3.10/site-packages/IPython/core/interactiveshell.py\", line 3132, in _run_cell\n",
      "    result = runner(coro)\n",
      "  File \"/Users/lorenzoturiano/Bocconi/ML-lab/ml_env/lib/python3.10/site-packages/IPython/core/async_helpers.py\", line 128, in _pseudo_sync_runner\n",
      "    coro.send(None)\n",
      "  File \"/Users/lorenzoturiano/Bocconi/ML-lab/ml_env/lib/python3.10/site-packages/IPython/core/interactiveshell.py\", line 3336, in run_cell_async\n",
      "    has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n",
      "  File \"/Users/lorenzoturiano/Bocconi/ML-lab/ml_env/lib/python3.10/site-packages/IPython/core/interactiveshell.py\", line 3519, in run_ast_nodes\n",
      "    if await self.run_code(code, result, async_=asy):\n",
      "  File \"/Users/lorenzoturiano/Bocconi/ML-lab/ml_env/lib/python3.10/site-packages/IPython/core/interactiveshell.py\", line 3579, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"/var/folders/xz/cl0qfkz56dl2q1b8vhqj2v180000gn/T/ipykernel_14233/3942993410.py\", line 2, in <module>\n",
      "    import torch\n",
      "  File \"/Users/lorenzoturiano/Bocconi/ML-lab/ml_env/lib/python3.10/site-packages/torch/__init__.py\", line 1477, in <module>\n",
      "    from .functional import *  # noqa: F403\n",
      "  File \"/Users/lorenzoturiano/Bocconi/ML-lab/ml_env/lib/python3.10/site-packages/torch/functional.py\", line 9, in <module>\n",
      "    import torch.nn.functional as F\n",
      "  File \"/Users/lorenzoturiano/Bocconi/ML-lab/ml_env/lib/python3.10/site-packages/torch/nn/__init__.py\", line 1, in <module>\n",
      "    from .modules import *  # noqa: F403\n",
      "  File \"/Users/lorenzoturiano/Bocconi/ML-lab/ml_env/lib/python3.10/site-packages/torch/nn/modules/__init__.py\", line 35, in <module>\n",
      "    from .transformer import TransformerEncoder, TransformerDecoder, \\\n",
      "  File \"/Users/lorenzoturiano/Bocconi/ML-lab/ml_env/lib/python3.10/site-packages/torch/nn/modules/transformer.py\", line 20, in <module>\n",
      "    device: torch.device = torch.device(torch._C._get_default_device()),  # torch.device('cpu'),\n",
      "/Users/lorenzoturiano/Bocconi/ML-lab/ml_env/lib/python3.10/site-packages/torch/nn/modules/transformer.py:20: UserWarning: Failed to initialize NumPy: _ARRAY_API not found (Triggered internally at /Users/runner/work/pytorch/pytorch/pytorch/torch/csrc/utils/tensor_numpy.cpp:84.)\n",
      "  device: torch.device = torch.device(torch._C._get_default_device()),  # torch.device('cpu'),\n"
     ]
    }
   ],
   "source": [
    "# PyTorch classifier\n",
    "import torch\n",
    "import torch.nn as nn\n",
    "from torch.utils.data import TensorDataset, DataLoader\n",
    "from sklearn.model_selection import train_test_split\n",
    "\n",
    "# Train-Test split\n",
    "X_train_np, X_test_np, y_train_np, y_test_np = train_test_split(\n",
    "    X.values, y, test_size=0.2, random_state=42, stratify=y\n",
    ")\n",
    "\n",
    "# Convert to tensors\n",
    "X_train = torch.tensor(X_train_np, dtype=torch.float32)\n",
    "y_train = torch.tensor(y_train_np, dtype=torch.long)\n",
    "X_test  = torch.tensor(X_test_np,  dtype=torch.float32)\n",
    "y_test  = torch.tensor(y_test_np,  dtype=torch.long)\n",
    "\n",
    "# Wrap in DataLoaders\n",
    "train_ds = TensorDataset(X_train, y_train)\n",
    "test_ds  = TensorDataset(X_test,  y_test)\n",
    "train_dl = DataLoader(train_ds, batch_size=32, shuffle=True)\n",
    "test_dl  = DataLoader(test_ds,  batch_size=32)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "28b070dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define the model: a small MLP\n",
    "class MLP(nn.Module):\n",
    "    def __init__(self, in_features, hidden=128, n_classes=4):\n",
    "        super().__init__()\n",
    "        self.net = nn.Sequential(\n",
    "            nn.Linear(in_features, hidden),\n",
    "            nn.ReLU(),\n",
    "            nn.Linear(hidden, n_classes)\n",
    "        )\n",
    "    def forward(self, x):\n",
    "        return self.net(x)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "f1c072a1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Epoch 1/50, loss: 91.3287\n",
      "Epoch 2/50, loss: 24.3034\n",
      "Epoch 3/50, loss: 6.1718\n",
      "Epoch 4/50, loss: 2.4856\n",
      "Epoch 5/50, loss: 0.9281\n",
      "Epoch 6/50, loss: 1.1944\n",
      "Epoch 7/50, loss: 0.6243\n",
      "Epoch 8/50, loss: 0.4053\n",
      "Epoch 9/50, loss: 0.2899\n",
      "Epoch 10/50, loss: 0.2742\n",
      "Epoch 11/50, loss: 0.2282\n",
      "Epoch 12/50, loss: 0.2178\n",
      "Epoch 13/50, loss: 0.4175\n",
      "Epoch 14/50, loss: 0.5181\n",
      "Epoch 15/50, loss: 0.2881\n",
      "Epoch 16/50, loss: 0.2617\n",
      "Epoch 17/50, loss: 0.2401\n",
      "Epoch 18/50, loss: 0.1896\n",
      "Epoch 19/50, loss: 0.2083\n",
      "Epoch 20/50, loss: 0.2814\n",
      "Epoch 21/50, loss: 0.2229\n",
      "Epoch 22/50, loss: 0.4478\n",
      "Epoch 23/50, loss: 0.4564\n",
      "Epoch 24/50, loss: 0.3685\n",
      "Epoch 25/50, loss: 0.3428\n",
      "Epoch 26/50, loss: 0.2773\n",
      "Epoch 27/50, loss: 0.2629\n",
      "Epoch 28/50, loss: 0.2645\n",
      "Epoch 29/50, loss: 0.2107\n",
      "Epoch 30/50, loss: 0.2252\n",
      "Epoch 31/50, loss: 0.2297\n",
      "Epoch 32/50, loss: 0.1862\n",
      "Epoch 33/50, loss: 0.1858\n",
      "Epoch 34/50, loss: 0.2133\n",
      "Epoch 35/50, loss: 0.1842\n",
      "Epoch 36/50, loss: 0.1547\n",
      "Epoch 37/50, loss: 0.1485\n",
      "Epoch 38/50, loss: 0.1992\n",
      "Epoch 39/50, loss: 0.2375\n",
      "Epoch 40/50, loss: 0.2162\n",
      "Epoch 41/50, loss: 0.1558\n",
      "Epoch 42/50, loss: 0.1645\n",
      "Epoch 43/50, loss: 0.1486\n",
      "Epoch 44/50, loss: 0.1625\n",
      "Epoch 45/50, loss: 0.1969\n",
      "Epoch 46/50, loss: 0.1688\n",
      "Epoch 47/50, loss: 0.1429\n",
      "Epoch 48/50, loss: 0.1800\n",
      "Epoch 49/50, loss: 0.2245\n",
      "Epoch 50/50, loss: 0.1527\n"
     ]
    }
   ],
   "source": [
    "device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')\n",
    "model  = MLP(in_features=X_train.shape[1]).to(device)\n",
    "loss_fn = nn.CrossEntropyLoss()\n",
    "opt     = torch.optim.Adam(model.parameters(), lr=1e-3)\n",
    "\n",
    "# 5) training loop\n",
    "n_epochs = 50\n",
    "for epoch in range(n_epochs):\n",
    "    model.train()\n",
    "    total_loss = 0\n",
    "    for xb, yb in train_dl:\n",
    "        xb, yb = xb.to(device), yb.to(device)\n",
    "        preds  = model(xb)\n",
    "        loss   = loss_fn(preds, yb)\n",
    "        opt.zero_grad()\n",
    "        loss.backward()\n",
    "        opt.step()\n",
    "        total_loss += loss.item()\n",
    "    avg_loss = total_loss / len(train_dl)\n",
    "    print(f\"Epoch {epoch+1}/{n_epochs}, loss: {avg_loss:.4f}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "241bef63",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Test accuracy: 78.61%\n"
     ]
    }
   ],
   "source": [
    "# Evaluation\n",
    "model.eval()\n",
    "correct = 0\n",
    "total   = 0\n",
    "with torch.no_grad():\n",
    "    for xb, yb in test_dl:\n",
    "        xb, yb = xb.to(device), yb.to(device)\n",
    "        preds  = model(xb).argmax(dim=1)\n",
    "        correct += (preds == yb).sum().item()\n",
    "        total   += yb.size(0)\n",
    "print(f\"Test accuracy: {correct/total:.2%}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cf1bc4ea",
   "metadata": {},
   "source": [
    "## What is a confusion matrix\n",
    "\n",
    "A confusion matrix is a table that shows how often a classifier’s predictions match the true labels.  \n",
    "- Each row is a true class  \n",
    "- Each column is a predicted class  \n",
    "- The cell at row i and column j is the count of samples whose true label is class i but the model predicted class j  \n",
    "\n",
    "## How we use it in this context\n",
    "\n",
    "We built a model to predict one of four mutation types: SNP, DNP, DEL, INS.  \n",
    "After running the model on the test set, we collect the true labels and the predicted labels, then build the confusion matrix. This tells us which mutation types the model gets right and which it confuses."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "4f1cbd04",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     SNP  DNP  DEL  INS\n",
      "SNP  119    0   39    2\n",
      "DNP    2    0    1    0\n",
      "DEL   14    0    5    0\n",
      "INS    5    0    0    0\n"
     ]
    }
   ],
   "source": [
    "from sklearn.metrics import confusion_matrix\n",
    "import pandas as pd\n",
    "\n",
    "# collect all true labels and predictions\n",
    "all_labels = []\n",
    "all_preds  = []\n",
    "\n",
    "model.eval()\n",
    "with torch.no_grad():\n",
    "    for xb, yb in test_dl:\n",
    "        xb, yb = xb.to(device), yb.to(device)\n",
    "        preds = model(xb).argmax(dim=1)\n",
    "        all_preds.extend(preds.cpu().tolist())\n",
    "        all_labels.extend(yb.cpu().tolist())\n",
    "\n",
    "# compute confusion matrix\n",
    "cm = confusion_matrix(all_labels, all_preds)\n",
    "\n",
    "# optional: pretty print with class names\n",
    "class_names = ['SNP', 'DNP', 'DEL', 'INS']\n",
    "cm_df = pd.DataFrame(cm, index=class_names, columns=class_names)\n",
    "print(cm_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "914a7c36",
   "metadata": {},
   "source": [
    "## The obtained confusion matrix\n",
    "\n",
    "Here is the table we got (rows = true label, columns = predicted label):\n",
    "\n",
    "| True \\ Predicted | SNP | DNP | DEL | INS |\n",
    "|------------------|-----|-----|-----|-----|\n",
    "| **SNP**          | 148 |   0 |   8 |   4 |\n",
    "| **DNP**          |   3 |   0 |   0 |   0 |\n",
    "| **DEL**          |  16 |   0 |   3 |   0 |\n",
    "| **INS**          |   5 |   0 |   0 |   0 |\n",
    "\n",
    "### What this tells us\n",
    "\n",
    "- **SNP**: 148 correct, 12 wrong (8 called DEL, 4 called INS)  \n",
    "- **DNP**: 0 correct, all 3 called SNP  \n",
    "- **DEL**: 3 correct, 16 called SNP  \n",
    "- **INS**: 0 correct, all 5 called SNP  \n",
    "\n",
    "The model is strong at spotting SNPs but it almost always labels DNP, DEL and INS as SNP.\n",
    "\n",
    "## Class distribution in the full dataset\n",
    "\n",
    "| Class | Samples |\n",
    "|-------|--------:|\n",
    "| SNP   |   1042  |\n",
    "| DEL   |    116  |\n",
    "| INS   |     34  |\n",
    "| DNP   |     26  |\n",
    "\n",
    "### Reasoning\n",
    "\n",
    "Because SNP makes up most of the data, the model sees many more SNP examples during training. This imbalance makes it easier for the model to learn SNP patterns and harder to learn the rarer classes (DEL, INS, DNP). As a result the model is biased toward predicting SNP. To fix this we can add more samples for the rare classes or use techniques like resampling or class weights to balance the training.  \n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "ml_env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.17"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
