{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ff2cb6cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a8ca2e5c",
   "metadata": {},
   "source": [
    "## TCGA\n",
    "\n",
    "## FPKM (Fragments Per Kilobase of transcript per Million)\n",
    "\n",
    "When you have paired-end RNA-seq data, you count **fragments** (read pairs) rather than individual reads. FPKM normalizes for sequencing depth and gene length just like RPKM, but uses fragments.\n",
    "\n",
    "1. **Sequencing depth**: samples sequenced to different depths get scaled so total fragment count differences don’t bias comparisons.  \n",
    "2. **Gene length**: longer transcripts produce more fragments just by being longer, so we divide out transcript length.\n",
    "\n",
    "The normalized \\((i, j)\\) FPKM value, where \\(i\\) is a gene and \\(j\\) is a sample, can be written two ways:\n",
    "\n",
    "$$\n",
    "\\mathrm{FPKM}_{i,j}\n",
    "=\n",
    "\\frac{f_{i,j}}\n",
    "     {\\,l_i \\cdot \\sum_{k} f_{k,j}\\,}\n",
    "\\times 10^6\n",
    "$$\n",
    "\n",
    "or equivalently\n",
    "\n",
    "$$\n",
    "\\mathrm{FPKM}_{i,j}\n",
    "=\n",
    "\\frac{f_{i,j} \\;\\times\\; 10^9}\n",
    "     {\\,N_j \\;\\times\\; L_i\\,}\n",
    "$$\n",
    "\n",
    "<!-- blank line above is important -->\n",
    "- **$f_{i,j}$** is the raw fragment count for gene \\(i\\) in sample \\(j\\).  \n",
    "- **$l_i$** is the length of gene \\(i\\) in kilobases (kb).  \n",
    "- **$\\sum_{k} f_{k,j}$** is the total number of fragments in sample \\(j\\).  \n",
    "- **$N_j$** is the total fragments sequenced in sample \\(j\\).  \n",
    "- **$L_i$** is the length of gene \\(i\\) in bases.  \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "67e0b620",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shape before merging duplicated rows: (56318, 1156)\n",
      "Shape after merging duplicated rows: (54353, 1156)\n"
     ]
    }
   ],
   "source": [
    "# mRNA seq\n",
    "# Expression Data with gene as index and cells/samples as columns\n",
    "data_mrna_seq_rpkm = pd.read_csv('tcga_data/data_mrna_seq_fpkm.txt',\n",
    "                    sep = '\\t',\n",
    "                    comment = '#')\n",
    "\n",
    "data_mrna_seq_rpkm.set_index('Hugo_Symbol',inplace=True)\n",
    "print(f\"Shape before merging duplicated rows: {data_mrna_seq_rpkm.shape}\")\n",
    "\n",
    "# Merge with mean duplicated rows\n",
    "data_mrna_seq_rpkm = data_mrna_seq_rpkm.groupby(data_mrna_seq_rpkm.index).mean()\n",
    "print(f\"Shape after merging duplicated rows: {data_mrna_seq_rpkm.shape}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a8916bbb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>22RV1_PROSTATE</th>\n",
       "      <th>2313287_STOMACH</th>\n",
       "      <th>253JBV_URINARY_TRACT</th>\n",
       "      <th>253J_URINARY_TRACT</th>\n",
       "      <th>42MGBA_CENTRAL_NERVOUS_SYSTEM</th>\n",
       "      <th>5637_URINARY_TRACT</th>\n",
       "      <th>59M_OVARY</th>\n",
       "      <th>639V_URINARY_TRACT</th>\n",
       "      <th>647V_URINARY_TRACT</th>\n",
       "      <th>697_HAEMATOPOIETIC_AND_LYMPHOID_TISSUE</th>\n",
       "      <th>...</th>\n",
       "      <th>UMUC16_URINARY_TRACT</th>\n",
       "      <th>UMUC4_URINARY_TRACT</th>\n",
       "      <th>UMUC5_URINARY_TRACT</th>\n",
       "      <th>UMUC6_URINARY_TRACT</th>\n",
       "      <th>UMUC7_URINARY_TRACT</th>\n",
       "      <th>UMUC9_URINARY_TRACT</th>\n",
       "      <th>UPCISCC152_UPPER_AERODIGESTIVE_TRACT</th>\n",
       "      <th>UW228_CENTRAL_NERVOUS_SYSTEM</th>\n",
       "      <th>Y79_AUTONOMIC_GANGLIA</th>\n",
       "      <th>YAMATO_SOFT_TISSUE</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CRIPTOP1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7SK</th>\n",
       "      <td>0.029785</td>\n",
       "      <td>0.03377</td>\n",
       "      <td>0.047608</td>\n",
       "      <td>0.02094</td>\n",
       "      <td>0.00430</td>\n",
       "      <td>0.017847</td>\n",
       "      <td>0.006893</td>\n",
       "      <td>0.00704</td>\n",
       "      <td>0.008042</td>\n",
       "      <td>0.148955</td>\n",
       "      <td>...</td>\n",
       "      <td>0.083175</td>\n",
       "      <td>0.02683</td>\n",
       "      <td>0.002032</td>\n",
       "      <td>0.024993</td>\n",
       "      <td>0.094365</td>\n",
       "      <td>0.393628</td>\n",
       "      <td>0.035012</td>\n",
       "      <td>0.002465</td>\n",
       "      <td>0.05875</td>\n",
       "      <td>0.067953</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>0.362300</td>\n",
       "      <td>0.00608</td>\n",
       "      <td>0.115170</td>\n",
       "      <td>0.33737</td>\n",
       "      <td>3.06452</td>\n",
       "      <td>0.013390</td>\n",
       "      <td>2.180160</td>\n",
       "      <td>2.24186</td>\n",
       "      <td>0.095400</td>\n",
       "      <td>3.698050</td>\n",
       "      <td>...</td>\n",
       "      <td>0.012700</td>\n",
       "      <td>0.09324</td>\n",
       "      <td>0.011340</td>\n",
       "      <td>0.060090</td>\n",
       "      <td>0.026380</td>\n",
       "      <td>0.019230</td>\n",
       "      <td>0.505580</td>\n",
       "      <td>0.578600</td>\n",
       "      <td>0.71115</td>\n",
       "      <td>0.849120</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1BG-AS1</th>\n",
       "      <td>4.815880</td>\n",
       "      <td>0.18959</td>\n",
       "      <td>0.751510</td>\n",
       "      <td>1.32578</td>\n",
       "      <td>9.15532</td>\n",
       "      <td>0.064590</td>\n",
       "      <td>1.619220</td>\n",
       "      <td>4.48351</td>\n",
       "      <td>0.631410</td>\n",
       "      <td>6.483210</td>\n",
       "      <td>...</td>\n",
       "      <td>0.371360</td>\n",
       "      <td>0.52954</td>\n",
       "      <td>0.187460</td>\n",
       "      <td>0.121010</td>\n",
       "      <td>0.359240</td>\n",
       "      <td>0.809130</td>\n",
       "      <td>2.750650</td>\n",
       "      <td>2.918060</td>\n",
       "      <td>2.06849</td>\n",
       "      <td>2.596400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1CF</th>\n",
       "      <td>5.654860</td>\n",
       "      <td>2.55482</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.12239</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.011540</td>\n",
       "      <td>0.00589</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 1156 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             22RV1_PROSTATE  2313287_STOMACH  253JBV_URINARY_TRACT  \\\n",
       "Hugo_Symbol                                                          \n",
       " CRIPTOP1          0.000000          0.00000              0.000000   \n",
       "7SK                0.029785          0.03377              0.047608   \n",
       "A1BG               0.362300          0.00608              0.115170   \n",
       "A1BG-AS1           4.815880          0.18959              0.751510   \n",
       "A1CF               5.654860          2.55482              0.000000   \n",
       "\n",
       "             253J_URINARY_TRACT  42MGBA_CENTRAL_NERVOUS_SYSTEM  \\\n",
       "Hugo_Symbol                                                      \n",
       " CRIPTOP1               0.00000                        0.00000   \n",
       "7SK                     0.02094                        0.00430   \n",
       "A1BG                    0.33737                        3.06452   \n",
       "A1BG-AS1                1.32578                        9.15532   \n",
       "A1CF                    0.00000                        0.00000   \n",
       "\n",
       "             5637_URINARY_TRACT  59M_OVARY  639V_URINARY_TRACT  \\\n",
       "Hugo_Symbol                                                      \n",
       " CRIPTOP1              0.000000   0.000000             0.00000   \n",
       "7SK                    0.017847   0.006893             0.00704   \n",
       "A1BG                   0.013390   2.180160             2.24186   \n",
       "A1BG-AS1               0.064590   1.619220             4.48351   \n",
       "A1CF                   0.000000   0.000000             0.00000   \n",
       "\n",
       "             647V_URINARY_TRACT  697_HAEMATOPOIETIC_AND_LYMPHOID_TISSUE  ...  \\\n",
       "Hugo_Symbol                                                              ...   \n",
       " CRIPTOP1              0.000000                                0.000000  ...   \n",
       "7SK                    0.008042                                0.148955  ...   \n",
       "A1BG                   0.095400                                3.698050  ...   \n",
       "A1BG-AS1               0.631410                                6.483210  ...   \n",
       "A1CF                   0.000000                                0.000000  ...   \n",
       "\n",
       "             UMUC16_URINARY_TRACT  UMUC4_URINARY_TRACT  UMUC5_URINARY_TRACT  \\\n",
       "Hugo_Symbol                                                                   \n",
       " CRIPTOP1                0.000000              0.00000             0.000000   \n",
       "7SK                      0.083175              0.02683             0.002032   \n",
       "A1BG                     0.012700              0.09324             0.011340   \n",
       "A1BG-AS1                 0.371360              0.52954             0.187460   \n",
       "A1CF                     0.000000              4.12239             0.000000   \n",
       "\n",
       "             UMUC6_URINARY_TRACT  UMUC7_URINARY_TRACT  UMUC9_URINARY_TRACT  \\\n",
       "Hugo_Symbol                                                                  \n",
       " CRIPTOP1               0.000000             0.000000             0.000000   \n",
       "7SK                     0.024993             0.094365             0.393628   \n",
       "A1BG                    0.060090             0.026380             0.019230   \n",
       "A1BG-AS1                0.121010             0.359240             0.809130   \n",
       "A1CF                    0.000000             0.000000             0.000000   \n",
       "\n",
       "             UPCISCC152_UPPER_AERODIGESTIVE_TRACT  \\\n",
       "Hugo_Symbol                                         \n",
       " CRIPTOP1                                0.000000   \n",
       "7SK                                      0.035012   \n",
       "A1BG                                     0.505580   \n",
       "A1BG-AS1                                 2.750650   \n",
       "A1CF                                     0.000000   \n",
       "\n",
       "             UW228_CENTRAL_NERVOUS_SYSTEM  Y79_AUTONOMIC_GANGLIA  \\\n",
       "Hugo_Symbol                                                        \n",
       " CRIPTOP1                        0.000000                0.00000   \n",
       "7SK                              0.002465                0.05875   \n",
       "A1BG                             0.578600                0.71115   \n",
       "A1BG-AS1                         2.918060                2.06849   \n",
       "A1CF                             0.011540                0.00589   \n",
       "\n",
       "             YAMATO_SOFT_TISSUE  \n",
       "Hugo_Symbol                      \n",
       " CRIPTOP1              0.000000  \n",
       "7SK                    0.067953  \n",
       "A1BG                   0.849120  \n",
       "A1BG-AS1               2.596400  \n",
       "A1CF                   0.000000  \n",
       "\n",
       "[5 rows x 1156 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_mrna_seq_rpkm.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "d509842a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>CRIPTOP1</th>\n",
       "      <th>7SK</th>\n",
       "      <th>A1BG</th>\n",
       "      <th>A1BG-AS1</th>\n",
       "      <th>A1CF</th>\n",
       "      <th>A2M</th>\n",
       "      <th>A2M-AS1</th>\n",
       "      <th>A2ML1</th>\n",
       "      <th>A2ML1-AS1</th>\n",
       "      <th>A2ML1-AS2</th>\n",
       "      <th>...</th>\n",
       "      <th>snoZ13_snr52</th>\n",
       "      <th>snoZ178</th>\n",
       "      <th>snoZ185</th>\n",
       "      <th>snoZ247</th>\n",
       "      <th>snoZ40</th>\n",
       "      <th>snoZ5</th>\n",
       "      <th>snoZ6</th>\n",
       "      <th>snosnR60_Z15</th>\n",
       "      <th>snosnR66</th>\n",
       "      <th>yR211F11.2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>22RV1_PROSTATE</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.029785</td>\n",
       "      <td>0.36230</td>\n",
       "      <td>4.81588</td>\n",
       "      <td>5.65486</td>\n",
       "      <td>1.98954</td>\n",
       "      <td>1.27348</td>\n",
       "      <td>0.01960</td>\n",
       "      <td>0.91070</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.341940</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0755</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2313287_STOMACH</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.033770</td>\n",
       "      <td>0.00608</td>\n",
       "      <td>0.18959</td>\n",
       "      <td>2.55482</td>\n",
       "      <td>0.00782</td>\n",
       "      <td>0.22274</td>\n",
       "      <td>0.01738</td>\n",
       "      <td>0.32801</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.239685</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>253JBV_URINARY_TRACT</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.047608</td>\n",
       "      <td>0.11517</td>\n",
       "      <td>0.75151</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.05044</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00544</td>\n",
       "      <td>0.60473</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>253J_URINARY_TRACT</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.020940</td>\n",
       "      <td>0.33737</td>\n",
       "      <td>1.32578</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.02878</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00547</td>\n",
       "      <td>0.46686</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42MGBA_CENTRAL_NERVOUS_SYSTEM</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.004300</td>\n",
       "      <td>3.06452</td>\n",
       "      <td>9.15532</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.02477</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.09900</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.000913</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 54353 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Hugo_Symbol                     CRIPTOP1       7SK     A1BG  A1BG-AS1  \\\n",
       "22RV1_PROSTATE                       0.0  0.029785  0.36230   4.81588   \n",
       "2313287_STOMACH                      0.0  0.033770  0.00608   0.18959   \n",
       "253JBV_URINARY_TRACT                 0.0  0.047608  0.11517   0.75151   \n",
       "253J_URINARY_TRACT                   0.0  0.020940  0.33737   1.32578   \n",
       "42MGBA_CENTRAL_NERVOUS_SYSTEM        0.0  0.004300  3.06452   9.15532   \n",
       "\n",
       "Hugo_Symbol                       A1CF      A2M  A2M-AS1    A2ML1  A2ML1-AS1  \\\n",
       "22RV1_PROSTATE                 5.65486  1.98954  1.27348  0.01960    0.91070   \n",
       "2313287_STOMACH                2.55482  0.00782  0.22274  0.01738    0.32801   \n",
       "253JBV_URINARY_TRACT           0.00000  0.05044  0.00000  0.00544    0.60473   \n",
       "253J_URINARY_TRACT             0.00000  0.02878  0.00000  0.00547    0.46686   \n",
       "42MGBA_CENTRAL_NERVOUS_SYSTEM  0.00000  0.02477  0.00000  0.00000    0.09900   \n",
       "\n",
       "Hugo_Symbol                    A2ML1-AS2  ...  snoZ13_snr52  snoZ178  snoZ185  \\\n",
       "22RV1_PROSTATE                       0.0  ...           0.0      0.0      0.0   \n",
       "2313287_STOMACH                      0.0  ...           0.0      0.0      0.0   \n",
       "253JBV_URINARY_TRACT                 0.0  ...           0.0      0.0      0.0   \n",
       "253J_URINARY_TRACT                   0.0  ...           0.0      0.0      0.0   \n",
       "42MGBA_CENTRAL_NERVOUS_SYSTEM        0.0  ...           0.0      0.0      0.0   \n",
       "\n",
       "Hugo_Symbol                    snoZ247  snoZ40  snoZ5     snoZ6  snosnR60_Z15  \\\n",
       "22RV1_PROSTATE                     0.0     0.0    0.0  3.341940           0.0   \n",
       "2313287_STOMACH                    0.0     0.0    0.0  3.239685           0.0   \n",
       "253JBV_URINARY_TRACT               0.0     0.0    0.0  0.000000           0.0   \n",
       "253J_URINARY_TRACT                 0.0     0.0    0.0  0.000000           0.0   \n",
       "42MGBA_CENTRAL_NERVOUS_SYSTEM      0.0     0.0    0.0  1.000913           0.0   \n",
       "\n",
       "Hugo_Symbol                    snosnR66  yR211F11.2  \n",
       "22RV1_PROSTATE                      0.0      0.0755  \n",
       "2313287_STOMACH                     0.0      0.0000  \n",
       "253JBV_URINARY_TRACT                0.0      0.0000  \n",
       "253J_URINARY_TRACT                  0.0      0.0000  \n",
       "42MGBA_CENTRAL_NERVOUS_SYSTEM       0.0      0.0000  \n",
       "\n",
       "[5 rows x 54353 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Transpose the data matrix to have a sample for each row\n",
    "data_t = data_mrna_seq_rpkm.T\n",
    "data_t.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "8301ef8b",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'pd' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[1], line 3\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;66;03m# Mutations matrix\u001b[39;00m\n\u001b[1;32m      2\u001b[0m \u001b[38;5;66;03m# it tells the type of mutation: WT (wild type) or else (actual mutation)\u001b[39;00m\n\u001b[0;32m----> 3\u001b[0m mutations\u001b[38;5;241m=\u001b[39m \u001b[43mpd\u001b[49m\u001b[38;5;241m.\u001b[39mread_csv(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mtcga_data/mutations.txt\u001b[39m\u001b[38;5;124m'\u001b[39m,\n\u001b[1;32m      4\u001b[0m                     sep \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;130;01m\\t\u001b[39;00m\u001b[38;5;124m'\u001b[39m,\n\u001b[1;32m      5\u001b[0m                     comment \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m#\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[1;32m      6\u001b[0m mutations\u001b[38;5;241m.\u001b[39mset_index(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mSAMPLE_ID\u001b[39m\u001b[38;5;124m'\u001b[39m,inplace\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m)\n\u001b[1;32m      7\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mMutations dataframe shape: \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mmutations\u001b[38;5;241m.\u001b[39mshape\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m)\n",
      "\u001b[0;31mNameError\u001b[0m: name 'pd' is not defined"
     ]
    }
   ],
   "source": [
    "# Mutations matrix\n",
    "# it tells the type of mutation: WT (wild type) or else (actual mutation)\n",
    "mutations= pd.read_csv('tcga_data/mutations.txt',\n",
    "                    sep = '\\t',\n",
    "                    comment = '#')\n",
    "mutations.set_index('SAMPLE_ID',inplace=True)\n",
    "print(f\"Mutations dataframe shape: {mutations.shape}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "269c9202",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>STUDY_ID</th>\n",
       "      <th>TP53</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMPLE_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SJRH30_SOFT_TISSUE</th>\n",
       "      <td>ccle_broad_2019</td>\n",
       "      <td>R280S R273C Y205C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FADU_UPPER_AERODIGESTIVE_TRACT</th>\n",
       "      <td>ccle_broad_2019</td>\n",
       "      <td>R248L X225_splice</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HH_HAEMATOPOIETIC_AND_LYMPHOID_TISSUE</th>\n",
       "      <td>ccle_broad_2019</td>\n",
       "      <td>X187_splice</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SKNMC_BONE</th>\n",
       "      <td>ccle_broad_2019</td>\n",
       "      <td>WT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SNU182_LIVER</th>\n",
       "      <td>ccle_broad_2019</td>\n",
       "      <td>S215I</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              STUDY_ID               TP53\n",
       "SAMPLE_ID                                                                \n",
       "SJRH30_SOFT_TISSUE                     ccle_broad_2019  R280S R273C Y205C\n",
       "FADU_UPPER_AERODIGESTIVE_TRACT         ccle_broad_2019  R248L X225_splice\n",
       "HH_HAEMATOPOIETIC_AND_LYMPHOID_TISSUE  ccle_broad_2019        X187_splice\n",
       "SKNMC_BONE                             ccle_broad_2019                 WT\n",
       "SNU182_LIVER                           ccle_broad_2019              S215I"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mutations.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "85a78da9",
   "metadata": {},
   "source": [
    "## Multi-class prediction: which classes to consider?\n",
    "We decided to use the `Variant_Type` classes (SNP, DNP, DEL, INS) as prediction targets for our classification task. The main reason is that we only have around 1,000+ samples in total, while each sample contains over 50,000+ gene expression levels. This creates a high-dimensional, low-sample-size scenario, which makes it important to simplify the prediction task to avoid overfitting and reduce noise.\n",
    "\n",
    "The `Variant_Classification` column contains many more categories, which would require more data to model accurately. In contrast, `Variant_Type` has only four well-defined mutation types (plus the Wild Type), making it more suitable for classification in our setting. These types are biologically meaningful and still capture key differences in mutation patterns. By focusing on `Variant_Type`, we ensure a more balanced and manageable classification problem given the size and structure of our dataset.\n",
    "\n",
    "The mutation types in `Variant_Type` are:\n",
    "\n",
    "- **SNP (Single Nucleotide Polymorphism)**: a single base pair in the DNA sequence is changed.\n",
    "- **DNP (Double Nucleotide Polymorphism)**: two adjacent base pairs are changed.\n",
    "- **DEL (Deletion)**: one or more nucleotides are deleted from the DNA sequence.\n",
    "- **INS (Insertion)**: one or more nucleotides are inserted into the DNA sequence."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ba6b31bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Data mutations matrix\n",
    "# Dataset containing mutation classes for all genes, contining only mutated samples\n",
    "data_mutations = pd.read_csv('tcga_data/data_mutations.txt',\n",
    "                    sep = '\\t',\n",
    "                    comment = '#')\n",
    "\n",
    "# Extract TP53 from all genes\n",
    "data_mutations = data_mutations[data_mutations['Hugo_Symbol'] == 'TP53']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "1d83c6e9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "9 different classes: ['Missense_Mutation' 'Nonsense_Mutation' 'Splice_Site' 'Silent'\n",
      " 'Splice_Region' 'Frame_Shift_Del' 'Frame_Shift_Ins' 'In_Frame_Del'\n",
      " 'In_Frame_Ins']\n"
     ]
    }
   ],
   "source": [
    "unique_labels_vc = data_mutations['Variant_Classification'].unique()\n",
    "print(f\"{len(unique_labels_vc)} different classes: {unique_labels_vc}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "7f29c365",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4 different classes: ['SNP' 'DNP' 'DEL' 'INS']\n"
     ]
    }
   ],
   "source": [
    "unique_labels_vt = data_mutations['Variant_Type'].unique()\n",
    "print(f\"{len(unique_labels_vt)} different classes: {unique_labels_vt}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "34cf6889",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remove unwanted information\n",
    "data_mutations = data_mutations[['Tumor_Sample_Barcode', 'Variant_Type']]\n",
    "data_mutations.set_index('Tumor_Sample_Barcode', inplace=True)\n",
    "\n",
    "# There are repetitions of my mutation type (Variant_Type)\n",
    "# if there is the same sample with different Variant_Type it should be removed\n",
    "variant_check = data_mutations.groupby(data_mutations.index)[\"Variant_Type\"].nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "748e4345",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Variant_Type</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tumor_Sample_Barcode</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>22RV1_PROSTATE</th>\n",
       "      <td>SNP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A431_SKIN</th>\n",
       "      <td>SNP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A4FUK_HAEMATOPOIETIC_AND_LYMPHOID_TISSUE</th>\n",
       "      <td>SNP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BICR16_UPPER_AERODIGESTIVE_TRACT</th>\n",
       "      <td>SNP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BICR78_UPPER_AERODIGESTIVE_TRACT</th>\n",
       "      <td>SNP</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         Variant_Type\n",
       "Tumor_Sample_Barcode                                 \n",
       "22RV1_PROSTATE                                    SNP\n",
       "A431_SKIN                                         SNP\n",
       "A4FUK_HAEMATOPOIETIC_AND_LYMPHOID_TISSUE          SNP\n",
       "BICR16_UPPER_AERODIGESTIVE_TRACT                  SNP\n",
       "BICR78_UPPER_AERODIGESTIVE_TRACT                  SNP"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_mutations.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "74dde40a",
   "metadata": {},
   "source": [
    "## Class balance check\n",
    "\n",
    "We check how many samples are in each class before training, because if one class is much larger the model can become biased toward it. A balanced dataset helps the model learn to recognize both classes equally."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "714a1aa5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'WT': 571, 'SNP': 835, 'DEL': 111, 'INS': 34, 'DNP': 19}\n"
     ]
    }
   ],
   "source": [
    "class_names = ['WT','SNP','DNP','DEL','INS']\n",
    "code = {\"WT\": 0, \"SNP\": 1, \"DNP\": 2, \"DEL\": 3, \"INS\": 4}\n",
    "variant_map = data_mutations['Variant_Type'].to_dict()\n",
    "\n",
    "class_counts = {\n",
    "    \"WT\": 0,\n",
    "    \"SNP\": 0,\n",
    "    \"DEL\": 0,\n",
    "    \"INS\": 0,\n",
    "    \"DNP\": 0\n",
    "}\n",
    "\n",
    "for sample_id, mut_type in mutations['TP53'].items():\n",
    "    if mut_type == 'WT':\n",
    "        class_counts[mut_type] += 1\n",
    "    else:\n",
    "        class_counts[variant_map[sample_id]] += 1\n",
    "\n",
    "print(class_counts)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a4798c4f",
   "metadata": {},
   "source": [
    "## Build target vector and data matrix for classifier\n",
    "To build an aligned data matrix X and a target vector y, we check if each sample in the mutations matrix has a correspondency in the original data matrix, and we discard the samples that don't have it."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "867c6ef2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Initial number of samples: 1156\n",
      "Number of samples discarded because of no correspondency in the data matrix: 30\n",
      "Final number of samples available: 1126\n"
     ]
    }
   ],
   "source": [
    "# Target vector\n",
    "y = []\n",
    "\n",
    "# Data matrix\n",
    "X_rows = []\n",
    "sample_names = []\n",
    "\n",
    "# Iterate over each mutation record\n",
    "for sample_id, mut_type in mutations['TP53'].items():\n",
    "    if (sample_id in data_t.index):\n",
    "        # Grab the full row from data_t and store it\n",
    "        X_rows.append(data_t.loc[sample_id].values)\n",
    "        sample_names.append(sample_id)\n",
    "        if mut_type == 'WT':\n",
    "            y.append(code['WT'])\n",
    "        else:\n",
    "            y.append(code[variant_map[sample_id]])\n",
    "\n",
    "# Build the new data matrix\n",
    "X = pd.DataFrame(\n",
    "    X_rows,\n",
    "    index=sample_names,\n",
    "    columns=data_t.columns\n",
    ")\n",
    "\n",
    "y = np.array(y)\n",
    "\n",
    "print(f\"Initial number of samples: {data_t.shape[0]}\")\n",
    "print(f\"Number of samples discarded because of no correspondency in the data matrix: {data_t.shape[0] - X.shape[0]}\")\n",
    "print(f\"Final number of samples available: {X.shape[0]}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "d9c47f9c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "X shape: (1126, 54353)\n",
      "y shape: (1126,)\n"
     ]
    }
   ],
   "source": [
    "# Print the shapes to make sure they are aligned\n",
    "print(f\"X shape: {X.shape}\")\n",
    "print(f\"y shape: {y.shape}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "97f30bde",
   "metadata": {},
   "source": [
    "## Train - Test set split"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "da468c7d",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.model_selection import train_test_split"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "76b0a91f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Train shape:\n",
      "\tX_train: (1013, 54353)\n",
      "\ty_train: (1013,)\n",
      "Test shape:\n",
      "\tX_test: (113, 54353)\n",
      "\ty_test: (113,)\n"
     ]
    }
   ],
   "source": [
    "# 90% train, 10% test, stratify to preserve class balance\n",
    "X_train_np, X_test_np, y_train_np, y_test_np = train_test_split(\n",
    "    X, y,\n",
    "    test_size=0.1,\n",
    "    random_state=10,\n",
    "    stratify=y\n",
    ")\n",
    "\n",
    "print(f\"Train shape:\\n\\tX_train: {X_train_np.shape}\\n\\ty_train: {y_train_np.shape}\")\n",
    "print(f\"Test shape:\\n\\tX_test: {X_test_np.shape}\\n\\ty_test: {y_test_np.shape}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "dd5dad50",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Train:\n",
      "\tWT samples: 340\n",
      "\tSNP samples: 560\n",
      "\tDNP samples: 10\n",
      "\tDEL samples: 81\n",
      "\tINS samples: 22\n",
      "Test:\n",
      "\tWT samples: 38\n",
      "\tSNP samples: 63\n",
      "\tDNP samples: 1\n",
      "\tDEL samples: 9\n",
      "\tINS samples: 2\n"
     ]
    }
   ],
   "source": [
    "# Check if Train - Test split is balanced in the number of classes\n",
    "inv_code = {v: k for k, v in code.items()}\n",
    "train_counts = {name: 0 for name in code.keys()}\n",
    "test_counts = {name: 0 for name in code.keys()}\n",
    "\n",
    "# Train\n",
    "for mut in y_train_np:\n",
    "    cls_name = inv_code[int(mut)]\n",
    "    train_counts[cls_name] += 1\n",
    "\n",
    "print(\"Train:\")\n",
    "for cls_name, count in train_counts.items():\n",
    "    print(f\"\\t{cls_name} samples: {count}\")\n",
    "\n",
    "# Test\n",
    "for mut in y_test_np:\n",
    "    cls_name = inv_code[int(mut)]\n",
    "    test_counts[cls_name] += 1\n",
    "\n",
    "print(\"Test:\")\n",
    "for cls_name, count in test_counts.items():\n",
    "    print(f\"\\t{cls_name} samples: {count}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "48156d5d",
   "metadata": {},
   "source": [
    "## Model selection and Training\n",
    "We try three different classifiers for the task.\n",
    "\n",
    "**Random forest** classifier: it is robust, accurate, and easy to use. It performs well with minimal preprocessing, handles high-dimensional data, and is resistant to overfitting thanks to its ensemble nature. Moreover, it can capture non-linear relationships and provides useful insights through feature importance scores, making it a strong and reliable choice for our classification task."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "99a1fd12",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.ensemble import RandomForestClassifier\n",
    "from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score\n",
    "from sklearn.model_selection import GridSearchCV"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "9937168c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to perform hyperparameter search for Random forest classifier\n",
    "def random_forest(X_train, y_train):\n",
    "    param_grid_rf = {\n",
    "        'n_estimators': [50, 100, 200],\n",
    "        'max_depth': [None, 10, 20, 30],\n",
    "        'min_samples_split': [2, 5, 10],\n",
    "        'min_samples_leaf': [1, 2, 4],\n",
    "        'max_features': ['log2', 'sqrt']\n",
    "    }\n",
    "\n",
    "    grid_search_rf = GridSearchCV(\n",
    "        estimator=RandomForestClassifier(random_state=42),\n",
    "        param_grid=param_grid_rf,\n",
    "        cv=5,\n",
    "        n_jobs=-1,\n",
    "        verbose=2\n",
    "    )\n",
    "\n",
    "    grid_search_rf.fit(X_train, y_train)\n",
    "\n",
    "    best_params_rf = grid_search_rf.best_params_\n",
    "    print(\"Best Random Forest Parameters:\", best_params_rf)\n",
    "\n",
    "    best_rf = grid_search_rf.best_estimator_\n",
    "    return best_rf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "9e1c7790",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>#sk-container-id-1 {\n",
       "  /* Definition of color scheme common for light and dark mode */\n",
       "  --sklearn-color-text: #000;\n",
       "  --sklearn-color-text-muted: #666;\n",
       "  --sklearn-color-line: gray;\n",
       "  /* Definition of color scheme for unfitted estimators */\n",
       "  --sklearn-color-unfitted-level-0: #fff5e6;\n",
       "  --sklearn-color-unfitted-level-1: #f6e4d2;\n",
       "  --sklearn-color-unfitted-level-2: #ffe0b3;\n",
       "  --sklearn-color-unfitted-level-3: chocolate;\n",
       "  /* Definition of color scheme for fitted estimators */\n",
       "  --sklearn-color-fitted-level-0: #f0f8ff;\n",
       "  --sklearn-color-fitted-level-1: #d4ebff;\n",
       "  --sklearn-color-fitted-level-2: #b3dbfd;\n",
       "  --sklearn-color-fitted-level-3: cornflowerblue;\n",
       "\n",
       "  /* Specific color for light theme */\n",
       "  --sklearn-color-text-on-default-background: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, black)));\n",
       "  --sklearn-color-background: var(--sg-background-color, var(--theme-background, var(--jp-layout-color0, white)));\n",
       "  --sklearn-color-border-box: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, black)));\n",
       "  --sklearn-color-icon: #696969;\n",
       "\n",
       "  @media (prefers-color-scheme: dark) {\n",
       "    /* Redefinition of color scheme for dark theme */\n",
       "    --sklearn-color-text-on-default-background: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, white)));\n",
       "    --sklearn-color-background: var(--sg-background-color, var(--theme-background, var(--jp-layout-color0, #111)));\n",
       "    --sklearn-color-border-box: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, white)));\n",
       "    --sklearn-color-icon: #878787;\n",
       "  }\n",
       "}\n",
       "\n",
       "#sk-container-id-1 {\n",
       "  color: var(--sklearn-color-text);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 pre {\n",
       "  padding: 0;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 input.sk-hidden--visually {\n",
       "  border: 0;\n",
       "  clip: rect(1px 1px 1px 1px);\n",
       "  clip: rect(1px, 1px, 1px, 1px);\n",
       "  height: 1px;\n",
       "  margin: -1px;\n",
       "  overflow: hidden;\n",
       "  padding: 0;\n",
       "  position: absolute;\n",
       "  width: 1px;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-dashed-wrapped {\n",
       "  border: 1px dashed var(--sklearn-color-line);\n",
       "  margin: 0 0.4em 0.5em 0.4em;\n",
       "  box-sizing: border-box;\n",
       "  padding-bottom: 0.4em;\n",
       "  background-color: var(--sklearn-color-background);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-container {\n",
       "  /* jupyter's `normalize.less` sets `[hidden] { display: none; }`\n",
       "     but bootstrap.min.css set `[hidden] { display: none !important; }`\n",
       "     so we also need the `!important` here to be able to override the\n",
       "     default hidden behavior on the sphinx rendered scikit-learn.org.\n",
       "     See: https://github.com/scikit-learn/scikit-learn/issues/21755 */\n",
       "  display: inline-block !important;\n",
       "  position: relative;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-text-repr-fallback {\n",
       "  display: none;\n",
       "}\n",
       "\n",
       "div.sk-parallel-item,\n",
       "div.sk-serial,\n",
       "div.sk-item {\n",
       "  /* draw centered vertical line to link estimators */\n",
       "  background-image: linear-gradient(var(--sklearn-color-text-on-default-background), var(--sklearn-color-text-on-default-background));\n",
       "  background-size: 2px 100%;\n",
       "  background-repeat: no-repeat;\n",
       "  background-position: center center;\n",
       "}\n",
       "\n",
       "/* Parallel-specific style estimator block */\n",
       "\n",
       "#sk-container-id-1 div.sk-parallel-item::after {\n",
       "  content: \"\";\n",
       "  width: 100%;\n",
       "  border-bottom: 2px solid var(--sklearn-color-text-on-default-background);\n",
       "  flex-grow: 1;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-parallel {\n",
       "  display: flex;\n",
       "  align-items: stretch;\n",
       "  justify-content: center;\n",
       "  background-color: var(--sklearn-color-background);\n",
       "  position: relative;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-parallel-item {\n",
       "  display: flex;\n",
       "  flex-direction: column;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-parallel-item:first-child::after {\n",
       "  align-self: flex-end;\n",
       "  width: 50%;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-parallel-item:last-child::after {\n",
       "  align-self: flex-start;\n",
       "  width: 50%;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-parallel-item:only-child::after {\n",
       "  width: 0;\n",
       "}\n",
       "\n",
       "/* Serial-specific style estimator block */\n",
       "\n",
       "#sk-container-id-1 div.sk-serial {\n",
       "  display: flex;\n",
       "  flex-direction: column;\n",
       "  align-items: center;\n",
       "  background-color: var(--sklearn-color-background);\n",
       "  padding-right: 1em;\n",
       "  padding-left: 1em;\n",
       "}\n",
       "\n",
       "\n",
       "/* Toggleable style: style used for estimator/Pipeline/ColumnTransformer box that is\n",
       "clickable and can be expanded/collapsed.\n",
       "- Pipeline and ColumnTransformer use this feature and define the default style\n",
       "- Estimators will overwrite some part of the style using the `sk-estimator` class\n",
       "*/\n",
       "\n",
       "/* Pipeline and ColumnTransformer style (default) */\n",
       "\n",
       "#sk-container-id-1 div.sk-toggleable {\n",
       "  /* Default theme specific background. It is overwritten whether we have a\n",
       "  specific estimator or a Pipeline/ColumnTransformer */\n",
       "  background-color: var(--sklearn-color-background);\n",
       "}\n",
       "\n",
       "/* Toggleable label */\n",
       "#sk-container-id-1 label.sk-toggleable__label {\n",
       "  cursor: pointer;\n",
       "  display: flex;\n",
       "  width: 100%;\n",
       "  margin-bottom: 0;\n",
       "  padding: 0.5em;\n",
       "  box-sizing: border-box;\n",
       "  text-align: center;\n",
       "  align-items: start;\n",
       "  justify-content: space-between;\n",
       "  gap: 0.5em;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 label.sk-toggleable__label .caption {\n",
       "  font-size: 0.6rem;\n",
       "  font-weight: lighter;\n",
       "  color: var(--sklearn-color-text-muted);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 label.sk-toggleable__label-arrow:before {\n",
       "  /* Arrow on the left of the label */\n",
       "  content: \"▸\";\n",
       "  float: left;\n",
       "  margin-right: 0.25em;\n",
       "  color: var(--sklearn-color-icon);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 label.sk-toggleable__label-arrow:hover:before {\n",
       "  color: var(--sklearn-color-text);\n",
       "}\n",
       "\n",
       "/* Toggleable content - dropdown */\n",
       "\n",
       "#sk-container-id-1 div.sk-toggleable__content {\n",
       "  max-height: 0;\n",
       "  max-width: 0;\n",
       "  overflow: hidden;\n",
       "  text-align: left;\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-0);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-toggleable__content.fitted {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-0);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-toggleable__content pre {\n",
       "  margin: 0.2em;\n",
       "  border-radius: 0.25em;\n",
       "  color: var(--sklearn-color-text);\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-0);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-toggleable__content.fitted pre {\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-0);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 input.sk-toggleable__control:checked~div.sk-toggleable__content {\n",
       "  /* Expand drop-down */\n",
       "  max-height: 200px;\n",
       "  max-width: 100%;\n",
       "  overflow: auto;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 input.sk-toggleable__control:checked~label.sk-toggleable__label-arrow:before {\n",
       "  content: \"▾\";\n",
       "}\n",
       "\n",
       "/* Pipeline/ColumnTransformer-specific style */\n",
       "\n",
       "#sk-container-id-1 div.sk-label input.sk-toggleable__control:checked~label.sk-toggleable__label {\n",
       "  color: var(--sklearn-color-text);\n",
       "  background-color: var(--sklearn-color-unfitted-level-2);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-label.fitted input.sk-toggleable__control:checked~label.sk-toggleable__label {\n",
       "  background-color: var(--sklearn-color-fitted-level-2);\n",
       "}\n",
       "\n",
       "/* Estimator-specific style */\n",
       "\n",
       "/* Colorize estimator box */\n",
       "#sk-container-id-1 div.sk-estimator input.sk-toggleable__control:checked~label.sk-toggleable__label {\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-2);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-estimator.fitted input.sk-toggleable__control:checked~label.sk-toggleable__label {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-2);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-label label.sk-toggleable__label,\n",
       "#sk-container-id-1 div.sk-label label {\n",
       "  /* The background is the default theme color */\n",
       "  color: var(--sklearn-color-text-on-default-background);\n",
       "}\n",
       "\n",
       "/* On hover, darken the color of the background */\n",
       "#sk-container-id-1 div.sk-label:hover label.sk-toggleable__label {\n",
       "  color: var(--sklearn-color-text);\n",
       "  background-color: var(--sklearn-color-unfitted-level-2);\n",
       "}\n",
       "\n",
       "/* Label box, darken color on hover, fitted */\n",
       "#sk-container-id-1 div.sk-label.fitted:hover label.sk-toggleable__label.fitted {\n",
       "  color: var(--sklearn-color-text);\n",
       "  background-color: var(--sklearn-color-fitted-level-2);\n",
       "}\n",
       "\n",
       "/* Estimator label */\n",
       "\n",
       "#sk-container-id-1 div.sk-label label {\n",
       "  font-family: monospace;\n",
       "  font-weight: bold;\n",
       "  display: inline-block;\n",
       "  line-height: 1.2em;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-label-container {\n",
       "  text-align: center;\n",
       "}\n",
       "\n",
       "/* Estimator-specific */\n",
       "#sk-container-id-1 div.sk-estimator {\n",
       "  font-family: monospace;\n",
       "  border: 1px dotted var(--sklearn-color-border-box);\n",
       "  border-radius: 0.25em;\n",
       "  box-sizing: border-box;\n",
       "  margin-bottom: 0.5em;\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-0);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-estimator.fitted {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-0);\n",
       "}\n",
       "\n",
       "/* on hover */\n",
       "#sk-container-id-1 div.sk-estimator:hover {\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-2);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-estimator.fitted:hover {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-2);\n",
       "}\n",
       "\n",
       "/* Specification for estimator info (e.g. \"i\" and \"?\") */\n",
       "\n",
       "/* Common style for \"i\" and \"?\" */\n",
       "\n",
       ".sk-estimator-doc-link,\n",
       "a:link.sk-estimator-doc-link,\n",
       "a:visited.sk-estimator-doc-link {\n",
       "  float: right;\n",
       "  font-size: smaller;\n",
       "  line-height: 1em;\n",
       "  font-family: monospace;\n",
       "  background-color: var(--sklearn-color-background);\n",
       "  border-radius: 1em;\n",
       "  height: 1em;\n",
       "  width: 1em;\n",
       "  text-decoration: none !important;\n",
       "  margin-left: 0.5em;\n",
       "  text-align: center;\n",
       "  /* unfitted */\n",
       "  border: var(--sklearn-color-unfitted-level-1) 1pt solid;\n",
       "  color: var(--sklearn-color-unfitted-level-1);\n",
       "}\n",
       "\n",
       ".sk-estimator-doc-link.fitted,\n",
       "a:link.sk-estimator-doc-link.fitted,\n",
       "a:visited.sk-estimator-doc-link.fitted {\n",
       "  /* fitted */\n",
       "  border: var(--sklearn-color-fitted-level-1) 1pt solid;\n",
       "  color: var(--sklearn-color-fitted-level-1);\n",
       "}\n",
       "\n",
       "/* On hover */\n",
       "div.sk-estimator:hover .sk-estimator-doc-link:hover,\n",
       ".sk-estimator-doc-link:hover,\n",
       "div.sk-label-container:hover .sk-estimator-doc-link:hover,\n",
       ".sk-estimator-doc-link:hover {\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-3);\n",
       "  color: var(--sklearn-color-background);\n",
       "  text-decoration: none;\n",
       "}\n",
       "\n",
       "div.sk-estimator.fitted:hover .sk-estimator-doc-link.fitted:hover,\n",
       ".sk-estimator-doc-link.fitted:hover,\n",
       "div.sk-label-container:hover .sk-estimator-doc-link.fitted:hover,\n",
       ".sk-estimator-doc-link.fitted:hover {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-3);\n",
       "  color: var(--sklearn-color-background);\n",
       "  text-decoration: none;\n",
       "}\n",
       "\n",
       "/* Span, style for the box shown on hovering the info icon */\n",
       ".sk-estimator-doc-link span {\n",
       "  display: none;\n",
       "  z-index: 9999;\n",
       "  position: relative;\n",
       "  font-weight: normal;\n",
       "  right: .2ex;\n",
       "  padding: .5ex;\n",
       "  margin: .5ex;\n",
       "  width: min-content;\n",
       "  min-width: 20ex;\n",
       "  max-width: 50ex;\n",
       "  color: var(--sklearn-color-text);\n",
       "  box-shadow: 2pt 2pt 4pt #999;\n",
       "  /* unfitted */\n",
       "  background: var(--sklearn-color-unfitted-level-0);\n",
       "  border: .5pt solid var(--sklearn-color-unfitted-level-3);\n",
       "}\n",
       "\n",
       ".sk-estimator-doc-link.fitted span {\n",
       "  /* fitted */\n",
       "  background: var(--sklearn-color-fitted-level-0);\n",
       "  border: var(--sklearn-color-fitted-level-3);\n",
       "}\n",
       "\n",
       ".sk-estimator-doc-link:hover span {\n",
       "  display: block;\n",
       "}\n",
       "\n",
       "/* \"?\"-specific style due to the `<a>` HTML tag */\n",
       "\n",
       "#sk-container-id-1 a.estimator_doc_link {\n",
       "  float: right;\n",
       "  font-size: 1rem;\n",
       "  line-height: 1em;\n",
       "  font-family: monospace;\n",
       "  background-color: var(--sklearn-color-background);\n",
       "  border-radius: 1rem;\n",
       "  height: 1rem;\n",
       "  width: 1rem;\n",
       "  text-decoration: none;\n",
       "  /* unfitted */\n",
       "  color: var(--sklearn-color-unfitted-level-1);\n",
       "  border: var(--sklearn-color-unfitted-level-1) 1pt solid;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 a.estimator_doc_link.fitted {\n",
       "  /* fitted */\n",
       "  border: var(--sklearn-color-fitted-level-1) 1pt solid;\n",
       "  color: var(--sklearn-color-fitted-level-1);\n",
       "}\n",
       "\n",
       "/* On hover */\n",
       "#sk-container-id-1 a.estimator_doc_link:hover {\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-3);\n",
       "  color: var(--sklearn-color-background);\n",
       "  text-decoration: none;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 a.estimator_doc_link.fitted:hover {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-3);\n",
       "}\n",
       "</style><div id=\"sk-container-id-1\" class=\"sk-top-container\"><div class=\"sk-text-repr-fallback\"><pre>RandomForestClassifier(max_depth=20, random_state=42)</pre><b>In a Jupyter environment, please rerun this cell to show the HTML representation or trust the notebook. <br />On GitHub, the HTML representation is unable to render, please try loading this page with nbviewer.org.</b></div><div class=\"sk-container\" hidden><div class=\"sk-item\"><div class=\"sk-estimator fitted sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-1\" type=\"checkbox\" checked><label for=\"sk-estimator-id-1\" class=\"sk-toggleable__label fitted sk-toggleable__label-arrow\"><div><div>RandomForestClassifier</div></div><div><a class=\"sk-estimator-doc-link fitted\" rel=\"noreferrer\" target=\"_blank\" href=\"https://scikit-learn.org/1.6/modules/generated/sklearn.ensemble.RandomForestClassifier.html\">?<span>Documentation for RandomForestClassifier</span></a><span class=\"sk-estimator-doc-link fitted\">i<span>Fitted</span></span></div></label><div class=\"sk-toggleable__content fitted\"><pre>RandomForestClassifier(max_depth=20, random_state=42)</pre></div> </div></div></div></div>"
      ],
      "text/plain": [
       "RandomForestClassifier(max_depth=20, random_state=42)"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Best Random Forest Parameters: {'max_depth': 20, 'max_features': 'sqrt', 'min_samples_leaf': 1, 'min_samples_split': 2, 'n_estimators': 100}\n",
    "\n",
    "# best_rf = random_forest(X_train_np, y_train_np) # UNCOMMENT HERE FOR HYPERPARAM SEARCH\n",
    "\n",
    "# Instantiate the model with the best parameters\n",
    "best_rf = RandomForestClassifier(\n",
    "    n_estimators=100,\n",
    "    max_depth=20,\n",
    "    max_features='sqrt',\n",
    "    min_samples_split=2,\n",
    "    min_samples_leaf=1,\n",
    "    random_state=42\n",
    ")\n",
    "\n",
    "# Train\n",
    "best_rf.fit(X_train_np, y_train_np)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "b4418614",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make predictions on test set\n",
    "y_pred_rf = best_rf.predict(X_test_np)\n",
    "\n",
    "# Evaluation\n",
    "accuracy_rf = accuracy_score(y_test_np, y_pred_rf)\n",
    "precision_rf = precision_score(y_test_np, y_pred_rf, average='weighted')\n",
    "recall_rf = recall_score(y_test_np, y_pred_rf, average='weighted')\n",
    "f1_rf = f1_score(y_test_np, y_pred_rf, average='weighted')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ad132628",
   "metadata": {},
   "source": [
    "**XGBoost** classifier: very good performance and efficiency. XGBoost is a powerful gradient boosting algorithm that handles missing values, supports regularization to reduce overfitting, and is optimized for speed. It works well with structured data and often outperforms other models, making it a strong choice for our problem."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "cf118d76",
   "metadata": {},
   "outputs": [],
   "source": [
    "from xgboost import XGBClassifier"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "e3fdcb21",
   "metadata": {},
   "outputs": [],
   "source": [
    "# XGBoost\n",
    "def xgboost(X_train, y_train):\n",
    "    param_grid_xgb_binary = {\n",
    "        'n_estimators': [100, 200],\n",
    "        'learning_rate': [0.01, 0.1],\n",
    "        'max_depth': [3, 5],\n",
    "        'subsample': [0.8, 1.0],\n",
    "        'colsample_bytree': [0.8, 1.0]\n",
    "    }\n",
    "\n",
    "\n",
    "    xgb = XGBClassifier(\n",
    "        objective='binary:logistic',\n",
    "        use_label_encoder=False,\n",
    "        eval_metric='logloss',\n",
    "        random_state=42\n",
    "    )\n",
    "\n",
    "\n",
    "    grid_search_xgb = GridSearchCV(\n",
    "        estimator=xgb,\n",
    "        param_grid=param_grid_xgb_binary,\n",
    "        cv=5,\n",
    "        n_jobs=-1,\n",
    "        verbose=2\n",
    "    )\n",
    "\n",
    "    grid_search_xgb.fit(X_train, y_train)\n",
    "\n",
    "    print(\"Best XGBoost Parameters:\", grid_search_xgb.best_params_)\n",
    "\n",
    "    best_xgb = grid_search_xgb.best_estimator_\n",
    "    return best_xgb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c490fc5a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Best XGBoost Parameters: {'colsample_bytree': 1.0, 'learning_rate': 0.1, 'max_depth': 4,\n",
    "#  'n_estimators': 100, 'subsample': 1.0}\n",
    "\n",
    "# best_xgb = xgboost(X_train_np, y_train_np) # UNCOMMENT HERE FOR HYPERPARAM SEARCH\n",
    "\n",
    "# Instantiate the model with the best parameters\n",
    "best_xgb = XGBClassifier(\n",
    "    colsample_bytree=1.0,\n",
    "    learning_rate=0.1,\n",
    "    max_depth=5,\n",
    "    n_estimators=100,\n",
    "    subsample=1.0,\n",
    "    random_state=42\n",
    ")\n",
    "\n",
    "# Train\n",
    "best_xgb.fit(X_train_np, y_train_np)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "144baec1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make predictions on test set\n",
    "y_pred_xgb = best_xgb.predict(X_test_np)\n",
    "\n",
    "# Evaluation\n",
    "accuracy_xgb = accuracy_score(y_test_np, y_pred_xgb)\n",
    "precision_xgb = precision_score(y_test_np, y_pred_xgb, average='weighted')\n",
    "recall_xgb = recall_score(y_test_np, y_pred_xgb, average='weighted')\n",
    "f1_xgb = f1_score(y_test_np, y_pred_xgb, average='weighted')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "15c8758f",
   "metadata": {},
   "source": [
    "**Multi-Layer Perceptron** classifier: neural networks are known for their ability to model complex, non-linear relationships in the data. By tuning its architecture and training parameters through grid search, we aimed to find the best configuration for our task. MLPs are flexible, work well with structured data, and can learn patterns that simpler models might miss, making them a solid option for classification problems."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0f536db8",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.neural_network import MLPClassifier"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0754189e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Neural Network (MLP)\n",
    "def neural_netw(X_train, y_train):\n",
    "    param_grid_nn = {\n",
    "        'hidden_layer_sizes': [(50,), (100,), (100, 50), (100, 100)],\n",
    "        'activation': ['relu', 'tanh'],\n",
    "        'solver': ['adam', 'sgd'],\n",
    "        'alpha': [0.0001, 0.001, 0.01],\n",
    "        'learning_rate': ['constant', 'adaptive']\n",
    "    }\n",
    "\n",
    "    grid_search_nn = GridSearchCV(\n",
    "        estimator=MLPClassifier(max_iter=1000, random_state=42),\n",
    "        param_grid=param_grid_nn,\n",
    "        cv=5,\n",
    "        n_jobs=-1,\n",
    "        verbose=2\n",
    "    )\n",
    "\n",
    "    grid_search_nn.fit(X_train, y_train)\n",
    "\n",
    "    best_params_nn = grid_search_nn.best_params_\n",
    "    print(\"Best Neural Network Parameters:\", best_params_nn)\n",
    "\n",
    "    best_nn = grid_search_nn.best_estimator_\n",
    "    return best_nn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0acc371b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Best Neural Network Parameters: {'activation': 'relu', 'alpha': 0.01, 'hidden_layer_sizes': (100, 50),\n",
    "# 'learning_rate': 'constant', 'solver': 'adam'}\n",
    "\n",
    "# best_nn = neural_netw(X_train_np, y_train_np) # UNCOMMENT HERE FOR HYPERPARAM SEARCH\n",
    "\n",
    "best_nn = MLPClassifier(\n",
    "    hidden_layer_sizes=(100, 50),\n",
    "    activation='relu',\n",
    "    solver='adam',\n",
    "    alpha=0.01,\n",
    "    learning_rate='constant',\n",
    "    max_iter=1000,\n",
    "    random_state=42\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f766d21f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make predictions on test set\n",
    "y_pred_nn = best_nn.predict(X_test_np)\n",
    "\n",
    "# Evaluation\n",
    "accuracy_nn = accuracy_score(y_test_np, y_pred_nn)\n",
    "precision_nn = precision_score(y_test_np, y_pred_nn, average='weighted')\n",
    "recall_nn = recall_score(y_test_np, y_pred_nn, average='weighted')\n",
    "f1_nn = f1_score(y_test_np, y_pred_nn, average='weighted')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "30351a72",
   "metadata": {},
   "source": [
    "## Evaluation metrics\n",
    "We evaluated all models using accuracy, precision, recall, and F1-score to provide a complete and balanced assessment of performance:\n",
    "\n",
    "- **Accuracy** gives an overall measure of how many predictions were correct.\n",
    "- **Precision (weighted)** helps understand how reliable the positive predictions are, while taking class imbalance into account.\n",
    "- **Recall (weighted)** shows how well the model identifies all relevant instances for each class.\n",
    "- **F1-score (weighted)** combines precision and recall into a single metric, balancing the two and adjusting for class distribution.\n",
    "\n",
    "These metrics are especially important in multi-class settings with possible class imbalance, as they give more insight than accuracy alone. The use of weighted averages ensures that all classes contribute proportionally to the final scores.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "17ef8d68",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Show metrics\n",
    "classifiers = [\n",
    "    (\"Random Forest\", accuracy_rf, precision_rf, recall_rf, f1_rf),\n",
    "    (\"XGBoost\", accuracy_xgb, precision_xgb, recall_xgb, f1_xgb),\n",
    "    (\"Neural Network (MLP)\", accuracy_nn, precision_nn, recall_nn, f1_nn)\n",
    "]\n",
    "\n",
    "for clf_name, accuracy, precision, recall, f1 in classifiers:\n",
    "    print(f\"{clf_name} Performance:\")\n",
    "    print(f\"  Accuracy: {accuracy:.4f} | Precision: {precision:.4f} | Recall: {recall:.4f} | F1 Score: {f1:.4f}\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "88d6690b",
   "metadata": {},
   "source": [
    "## Confusion Matrices\n",
    "We computed the confusion matrix for each model (Random Forest, XGBoost, and Neural Network) to analyze their classification performance in more detail. The confusion matrix shows how often each class was correctly or incorrectly predicted. It helps us identify specific types of errors, such as which classes are most often confused with one another.\n",
    "\n",
    "This is especially useful in multi-class classification tasks, where accuracy alone might not reveal which classes are harder to distinguish. By comparing the confusion matrices, we can better understand the strengths and weaknesses of each model and identify patterns of misclassification."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "53943b6e",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.metrics import confusion_matrix\n",
    "\n",
    "cm_rf = confusion_matrix(y_test_np, y_pred_rf)\n",
    "cm_xgb = confusion_matrix(y_test_np, y_pred_xgb)\n",
    "cm_nn = confusion_matrix(y_test_np, y_pred_nn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "68cc5169",
   "metadata": {},
   "outputs": [],
   "source": [
    "classifier_names = [\n",
    "    'Random Forest', 'XGBoost', 'Neural Network'\n",
    "]\n",
    "\n",
    "fig, axes = plt.subplots(1, 3, figsize=(10, 8))\n",
    "fig.suptitle('Confusion Matrices', fontsize=16)\n",
    "\n",
    "axes = axes.ravel()\n",
    "\n",
    "for i, (cm, name) in enumerate(zip(\n",
    "    [cm_rf, cm_xgb, cm_nn],\n",
    "    classifier_names\n",
    ")):\n",
    "    sns.heatmap(cm, annot=True, fmt='d', cmap='Blues', ax=axes[i], cbar=False)\n",
    "    axes[i].set_title(name)\n",
    "    axes[i].set_xlabel('Predicted')\n",
    "    axes[i].set_ylabel('Actual')\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "ml_env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.17"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
