{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "72154775",
   "metadata": {},
   "source": [
    "## Description\n",
    "\n",
    "This CCLE dataset is made by performing the usual bulk RNA-seq technique. It was then normalized using RPKM (Reads Per Kilobase of transcript per Million) to reduce variance in order to make samples comparable.\n",
    "\n",
    "The RPKM normalization solves two problems that are created during sequencing:\n",
    "\n",
    "1. **Sequencing depth**: it occurs when samples are sequenced differently, some samples may have been sequenced more or less than others. This effect does not reflect biological status, so we should correct for that.  \n",
    "2. **Gene length**: the gene length tells us how easy it is for that gene to be detected by sequencing machines. On one hand, longer genes are easier to detect, so their count will be bigger. On the other hand, smaller genes are more difficult. Normalizations on gene length take this into account.\n",
    "\n",
    "The normalised \\((i, j)\\) RPKM value where \\(i\\) is a gene and \\(j\\) is a sample:\n",
    "\n",
    "\\[\n",
    "\\mathrm{RPKM}_{i,j} \\;=\\; \\frac{x_{i,j}}{l_i \\,\\cdot\\, \\sum_j x_{i,j}} \\;\\times\\; 10^6\n",
    "\\]\n",
    "\n",
    "Where \\(x_{i,j}\\) is the raw count, \\(l_i\\) is gene length in kilobases (kb), and the denominator is total reads in sample \\(j\\).\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "3403de5a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "67e0b620",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ---------mRNA---------\n",
    "# Description:\n",
    "# Expression Data with gene as index and cells/samples as columns\n",
    "data_mrna_seq_rpkm = pd.read_csv('tcga_data/data_mrna_seq_fpkm.txt',\n",
    "                    sep = '\\t',\n",
    "                    comment = '#')\n",
    "\n",
    "data_mrna_seq_rpkm.set_index('Hugo_Symbol',inplace=True)\n",
    "\n",
    "# Merge with mean duplicated rows\n",
    "data_mrna_seq_rpkm = data_mrna_seq_rpkm.groupby(data_mrna_seq_rpkm.index).mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "a8916bbb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SP89389</th>\n",
       "      <th>SP21193</th>\n",
       "      <th>SP13206</th>\n",
       "      <th>SP103623</th>\n",
       "      <th>SP32742</th>\n",
       "      <th>SP111095</th>\n",
       "      <th>SP8394</th>\n",
       "      <th>SP87446</th>\n",
       "      <th>SP36586</th>\n",
       "      <th>SP123902</th>\n",
       "      <th>...</th>\n",
       "      <th>SP15656</th>\n",
       "      <th>SP123888</th>\n",
       "      <th>SP59420</th>\n",
       "      <th>SP116679</th>\n",
       "      <th>SP1377</th>\n",
       "      <th>SP16269</th>\n",
       "      <th>SP122676</th>\n",
       "      <th>SP88776</th>\n",
       "      <th>SP64546</th>\n",
       "      <th>SP21057</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CRIPTOP1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>0.409799</td>\n",
       "      <td>0.022891</td>\n",
       "      <td>0.082842</td>\n",
       "      <td>0.075852</td>\n",
       "      <td>0.149032</td>\n",
       "      <td>0.023175</td>\n",
       "      <td>0.119738</td>\n",
       "      <td>0.112135</td>\n",
       "      <td>0.059707</td>\n",
       "      <td>0.021957</td>\n",
       "      <td>...</td>\n",
       "      <td>0.587783</td>\n",
       "      <td>0.029553</td>\n",
       "      <td>0.229903</td>\n",
       "      <td>0.315114</td>\n",
       "      <td>0.034723</td>\n",
       "      <td>0.656322</td>\n",
       "      <td>0.476527</td>\n",
       "      <td>0.079750</td>\n",
       "      <td>0.091196</td>\n",
       "      <td>0.013030</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1BG-AS1</th>\n",
       "      <td>1.567399</td>\n",
       "      <td>0.196993</td>\n",
       "      <td>0.376265</td>\n",
       "      <td>0.271986</td>\n",
       "      <td>1.035252</td>\n",
       "      <td>0.132957</td>\n",
       "      <td>1.357942</td>\n",
       "      <td>0.753626</td>\n",
       "      <td>0.102765</td>\n",
       "      <td>0.110224</td>\n",
       "      <td>...</td>\n",
       "      <td>3.424922</td>\n",
       "      <td>0.378225</td>\n",
       "      <td>1.741374</td>\n",
       "      <td>1.946055</td>\n",
       "      <td>0.946255</td>\n",
       "      <td>3.781337</td>\n",
       "      <td>1.074688</td>\n",
       "      <td>0.554893</td>\n",
       "      <td>0.433442</td>\n",
       "      <td>0.074755</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1CF</th>\n",
       "      <td>0.043229</td>\n",
       "      <td>0.885682</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.715912</td>\n",
       "      <td>0.010971</td>\n",
       "      <td>2.296044</td>\n",
       "      <td>0.027879</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.656342</td>\n",
       "      <td>0.016029</td>\n",
       "      <td>...</td>\n",
       "      <td>0.049806</td>\n",
       "      <td>0.004742</td>\n",
       "      <td>0.008621</td>\n",
       "      <td>0.013397</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.012708</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000708</td>\n",
       "      <td>0.001492</td>\n",
       "      <td>3.038471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>30.816782</td>\n",
       "      <td>57.959083</td>\n",
       "      <td>37.798822</td>\n",
       "      <td>224.339366</td>\n",
       "      <td>23.396122</td>\n",
       "      <td>14.215019</td>\n",
       "      <td>53.002411</td>\n",
       "      <td>47.629206</td>\n",
       "      <td>51.087998</td>\n",
       "      <td>39.520998</td>\n",
       "      <td>...</td>\n",
       "      <td>0.023052</td>\n",
       "      <td>33.442775</td>\n",
       "      <td>53.716981</td>\n",
       "      <td>80.680141</td>\n",
       "      <td>23.491400</td>\n",
       "      <td>0.048657</td>\n",
       "      <td>121.201090</td>\n",
       "      <td>41.204146</td>\n",
       "      <td>33.427796</td>\n",
       "      <td>28.183744</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 1210 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               SP89389    SP21193    SP13206    SP103623    SP32742  \\\n",
       "Hugo_Symbol                                                           \n",
       " CRIPTOP1     0.000000   0.000000   0.000000    0.000000   0.000000   \n",
       "A1BG          0.409799   0.022891   0.082842    0.075852   0.149032   \n",
       "A1BG-AS1      1.567399   0.196993   0.376265    0.271986   1.035252   \n",
       "A1CF          0.043229   0.885682   0.000000    0.715912   0.010971   \n",
       "A2M          30.816782  57.959083  37.798822  224.339366  23.396122   \n",
       "\n",
       "              SP111095     SP8394    SP87446    SP36586   SP123902  ...  \\\n",
       "Hugo_Symbol                                                         ...   \n",
       " CRIPTOP1     0.000000   0.000000   0.000000   0.000000   0.000000  ...   \n",
       "A1BG          0.023175   0.119738   0.112135   0.059707   0.021957  ...   \n",
       "A1BG-AS1      0.132957   1.357942   0.753626   0.102765   0.110224  ...   \n",
       "A1CF          2.296044   0.027879   0.000000   1.656342   0.016029  ...   \n",
       "A2M          14.215019  53.002411  47.629206  51.087998  39.520998  ...   \n",
       "\n",
       "              SP15656   SP123888    SP59420   SP116679     SP1377   SP16269  \\\n",
       "Hugo_Symbol                                                                   \n",
       " CRIPTOP1    0.000000   0.000000   0.000000   0.000000   0.000000  0.000000   \n",
       "A1BG         0.587783   0.029553   0.229903   0.315114   0.034723  0.656322   \n",
       "A1BG-AS1     3.424922   0.378225   1.741374   1.946055   0.946255  3.781337   \n",
       "A1CF         0.049806   0.004742   0.008621   0.013397   0.000000  0.012708   \n",
       "A2M          0.023052  33.442775  53.716981  80.680141  23.491400  0.048657   \n",
       "\n",
       "               SP122676    SP88776    SP64546    SP21057  \n",
       "Hugo_Symbol                                               \n",
       " CRIPTOP1      0.000000   0.000000   0.000000   0.000000  \n",
       "A1BG           0.476527   0.079750   0.091196   0.013030  \n",
       "A1BG-AS1       1.074688   0.554893   0.433442   0.074755  \n",
       "A1CF           0.000000   0.000708   0.001492   3.038471  \n",
       "A2M          121.201090  41.204146  33.427796  28.183744  \n",
       "\n",
       "[5 rows x 1210 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_mrna_seq_rpkm.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "8301ef8b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#--------Mut query----------\n",
    "# Description:\n",
    "# Data telling  which sample is mutated and not-mutated\n",
    "mutations= pd.read_csv('tcga_data/mutations.txt',\n",
    "                    sep = '\\t',\n",
    "                    comment = '#')\n",
    "mutations.set_index('SAMPLE_ID',inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "072c2041",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>STUDY_ID</th>\n",
       "      <th>TP53</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMPLE_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SP107436</th>\n",
       "      <td>pancan_pcawg_2020</td>\n",
       "      <td>WT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SP107435</th>\n",
       "      <td>pancan_pcawg_2020</td>\n",
       "      <td>WT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SP107407</th>\n",
       "      <td>pancan_pcawg_2020</td>\n",
       "      <td>WT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SP107406</th>\n",
       "      <td>pancan_pcawg_2020</td>\n",
       "      <td>WT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SP107405</th>\n",
       "      <td>pancan_pcawg_2020</td>\n",
       "      <td>WT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    STUDY_ID TP53\n",
       "SAMPLE_ID                        \n",
       "SP107436   pancan_pcawg_2020   WT\n",
       "SP107435   pancan_pcawg_2020   WT\n",
       "SP107407   pancan_pcawg_2020   WT\n",
       "SP107406   pancan_pcawg_2020   WT\n",
       "SP107405   pancan_pcawg_2020   WT"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mutations.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "006ae621",
   "metadata": {},
   "source": [
    "## Why predict on just the five Variant_Type classes\n",
    "\n",
    "We focus on SNP, DNP, ONP, INS, and DEL because:\n",
    "\n",
    "1. Biological clarity: these five labels describe the fundamental mutation mechanism, single-base or multi-base changes, insertions, and deletions, so the model learns clear patterns.  \n",
    "2. Balanced data: each Variant_Type occurs often enough to give the model enough examples, while the detailed Variant_Classification labels are very uneven and would leave some classes too small to learn.  \n",
    "3. Reduced complexity: Variant_Classification depends on gene structure and reading frame (for example a SNP can be silent or missense depending on codon), which our sequence‐only model cannot infer without extra annotation.  \n",
    "4. Modular workflow: once the model tags a variant as INS or DEL, we can apply separate rules or a second model to predict functional impact, keeping each step simpler and more reliable.  \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "e215f51c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ------Mut ALL-----------\n",
    "# Dataset containing mutation classes for all genes, it only contains mutated samples\n",
    "data_mutations = pd.read_csv('tcga_data/data_mutations.txt',\n",
    "                    sep = '\\t',\n",
    "                    comment = '#')\n",
    "\n",
    "# Extract TP53 from all genes\n",
    "data_mutations = data_mutations[data_mutations['Hugo_Symbol'] == 'TP53']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "c240f890",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['SNP' 'DEL' 'INS' 'DNP']\n"
     ]
    }
   ],
   "source": [
    "unique_labels = data_mutations['Variant_Type'].unique()\n",
    "print(unique_labels)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "7f29c365",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Nonsense_Mutation' 'Missense_Mutation' 'In_Frame_Del' 'Frame_Shift_Del'\n",
      " 'Splice_Site' 'Frame_Shift_Ins' 'Splice_Region' 'In_Frame_Ins' 'Silent']\n"
     ]
    }
   ],
   "source": [
    "unique_labels = data_mutations['Variant_Classification'].unique()\n",
    "print(unique_labels)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "34cf6889",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remove unwanted information\n",
    "data_mutations = data_mutations[['Tumor_Sample_Barcode', 'Variant_Type']]\n",
    "data_mutations.set_index('Tumor_Sample_Barcode', inplace=True)\n",
    "# There are repetitions of my mutation type (Variant_Type)\n",
    "# if there is the same sample with different Variant_Type it should be removed\n",
    "variant_check = data_mutations.groupby(data_mutations.index)[\"Variant_Type\"].nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "748e4345",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Variant_Type</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tumor_Sample_Barcode</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SP101724</th>\n",
       "      <td>SNP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SP22031</th>\n",
       "      <td>SNP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SP59388</th>\n",
       "      <td>SNP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SP94588</th>\n",
       "      <td>DEL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SP7692</th>\n",
       "      <td>SNP</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Variant_Type\n",
       "Tumor_Sample_Barcode             \n",
       "SP101724                      SNP\n",
       "SP22031                       SNP\n",
       "SP59388                       SNP\n",
       "SP94588                       DEL\n",
       "SP7692                        SNP"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_mutations.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "714a1aa5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SNP: 772\n",
      "DEL: 115\n",
      "INS: 43\n",
      "DNP: 11\n"
     ]
    }
   ],
   "source": [
    "# Count how many mutations are in each Variant_Type\n",
    "counts = data_mutations['Variant_Type'].value_counts()\n",
    "\n",
    "# Print them one per line\n",
    "for variant_type, n in counts.items():\n",
    "    print(f\"{variant_type}: {n}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "965665f0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>CRIPTOP1</th>\n",
       "      <th>A1BG</th>\n",
       "      <th>A1BG-AS1</th>\n",
       "      <th>A1CF</th>\n",
       "      <th>A2M</th>\n",
       "      <th>A2M-AS1</th>\n",
       "      <th>A2ML1</th>\n",
       "      <th>A2ML1-AS1</th>\n",
       "      <th>A2ML1-AS2</th>\n",
       "      <th>A2MP1</th>\n",
       "      <th>...</th>\n",
       "      <th>snoZ178</th>\n",
       "      <th>snoZ185</th>\n",
       "      <th>snoZ247</th>\n",
       "      <th>snoZ278</th>\n",
       "      <th>snoZ40</th>\n",
       "      <th>snoZ5</th>\n",
       "      <th>snoZ6</th>\n",
       "      <th>snosnR60_Z15</th>\n",
       "      <th>snosnR66</th>\n",
       "      <th>yR211F11.2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SP89389</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.409799</td>\n",
       "      <td>1.567399</td>\n",
       "      <td>0.043229</td>\n",
       "      <td>30.816782</td>\n",
       "      <td>0.293354</td>\n",
       "      <td>0.312641</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.146655</td>\n",
       "      <td>0.024311</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>12.87572</td>\n",
       "      <td>0.085771</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SP21193</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.022891</td>\n",
       "      <td>0.196993</td>\n",
       "      <td>0.885682</td>\n",
       "      <td>57.959083</td>\n",
       "      <td>0.125789</td>\n",
       "      <td>0.006278</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.059391</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SP13206</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.082842</td>\n",
       "      <td>0.376265</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>37.798822</td>\n",
       "      <td>0.056904</td>\n",
       "      <td>1.130378</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.035823</td>\n",
       "      <td>0.026723</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SP103623</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.075852</td>\n",
       "      <td>0.271986</td>\n",
       "      <td>0.715912</td>\n",
       "      <td>224.339366</td>\n",
       "      <td>1.493611</td>\n",
       "      <td>0.114422</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.179434</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SP32742</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.149032</td>\n",
       "      <td>1.035252</td>\n",
       "      <td>0.010971</td>\n",
       "      <td>23.396122</td>\n",
       "      <td>0.235518</td>\n",
       "      <td>29.877306</td>\n",
       "      <td>0.051798</td>\n",
       "      <td>0.030327</td>\n",
       "      <td>0.045247</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.156084</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.020183</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 55851 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Hugo_Symbol   CRIPTOP1      A1BG  A1BG-AS1      A1CF         A2M   A2M-AS1  \\\n",
       "SP89389            0.0  0.409799  1.567399  0.043229   30.816782  0.293354   \n",
       "SP21193            0.0  0.022891  0.196993  0.885682   57.959083  0.125789   \n",
       "SP13206            0.0  0.082842  0.376265  0.000000   37.798822  0.056904   \n",
       "SP103623           0.0  0.075852  0.271986  0.715912  224.339366  1.493611   \n",
       "SP32742            0.0  0.149032  1.035252  0.010971   23.396122  0.235518   \n",
       "\n",
       "Hugo_Symbol      A2ML1  A2ML1-AS1  A2ML1-AS2     A2MP1  ...  snoZ178  snoZ185  \\\n",
       "SP89389       0.312641   0.000000   0.146655  0.024311  ...      0.0      0.0   \n",
       "SP21193       0.006278   0.000000   0.059391  0.000000  ...      0.0      0.0   \n",
       "SP13206       1.130378   0.000000   0.035823  0.026723  ...      0.0      0.0   \n",
       "SP103623      0.114422   0.000000   0.000000  0.179434  ...      0.0      0.0   \n",
       "SP32742      29.877306   0.051798   0.030327  0.045247  ...      0.0      0.0   \n",
       "\n",
       "Hugo_Symbol  snoZ247  snoZ278  snoZ40     snoZ5     snoZ6  snosnR60_Z15  \\\n",
       "SP89389          0.0      0.0     0.0  12.87572  0.085771           0.0   \n",
       "SP21193          0.0      0.0     0.0   0.00000  0.000000           0.0   \n",
       "SP13206          0.0      0.0     0.0   0.00000  0.000000           0.0   \n",
       "SP103623         0.0      0.0     0.0   0.00000  0.000000           0.0   \n",
       "SP32742          0.0      0.0     0.0   0.00000  0.156084           0.0   \n",
       "\n",
       "Hugo_Symbol  snosnR66  yR211F11.2  \n",
       "SP89389           0.0    0.000000  \n",
       "SP21193           0.0    0.000000  \n",
       "SP13206           0.0    0.000000  \n",
       "SP103623          0.0    0.000000  \n",
       "SP32742           0.0    0.020183  \n",
       "\n",
       "[5 rows x 55851 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_t = data_mrna_seq_rpkm.T\n",
    "\n",
    "data_t.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "cad8e5cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "code = {\"SNP\": 1, \"DNP\": 2, \"DEL\": 3, \"INS\": 4}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "a01e0551",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of samples discarded: 510\n"
     ]
    }
   ],
   "source": [
    "# Build target vector y\n",
    "y = []\n",
    "\n",
    "# prepare lists to collect rows and their sample name\n",
    "X_rows = []\n",
    "sample_names = []\n",
    "c = 0\n",
    "\n",
    "# iterate over each mutation record\n",
    "for bc, mut in data_mutations.iterrows():\n",
    "    # check if this barcode is in data_t’s index\n",
    "    if bc in data_t.index:\n",
    "        # grab the full row from data_t and store it\n",
    "        X_rows.append(data_t.loc[bc].values)\n",
    "        y.append(code[mut['Variant_Type']])\n",
    "        sample_names.append(bc)\n",
    "    else:\n",
    "        c += 1\n",
    "\n",
    "print(f\"Number of samples discarded: {c}\")\n",
    "\n",
    "# build a new DataFrame X from the collected rows\n",
    "X = pd.DataFrame(\n",
    "    X_rows,\n",
    "    index=sample_names,\n",
    "    columns=data_t.columns\n",
    ")\n",
    "\n",
    "y = np.array(y)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "d9c47f9c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "X shape: (431, 55851)\n",
      "y shape: (431,)\n"
     ]
    }
   ],
   "source": [
    "print(f\"X shape: {X.shape}\")\n",
    "print(f\"y shape: {y.shape}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "97f30bde",
   "metadata": {},
   "source": [
    "**Step 2: Train - Test split (80% - 20%)**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "da468c7d",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.model_selection import train_test_split"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "76b0a91f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Train shape:\n",
      "\tX_train: (344, 55851)\n",
      "\ty_train: (344,)\n",
      "Test shape:\n",
      "\tX_test: (87, 55851)\n",
      "\ty_test: (87,)\n"
     ]
    }
   ],
   "source": [
    "# 80% train, 20% test, stratify to preserve class balance\n",
    "X_train, X_test, y_train, y_test = train_test_split(\n",
    "    X, y,\n",
    "    test_size=0.2,\n",
    "    random_state=42,\n",
    "    stratify=y\n",
    ")\n",
    "\n",
    "print(f\"Train shape:\\n\\tX_train: {X_train.shape}\\n\\ty_train: {y_train.shape}\")\n",
    "print(f\"Test shape:\\n\\tX_test: {X_test.shape}\\n\\ty_test: {y_test.shape}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "48156d5d",
   "metadata": {},
   "source": [
    "**Step 3: Model selection and Training**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "99a1fd12",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.ensemble import RandomForestClassifier\n",
    "from sklearn.metrics import classification_report, confusion_matrix, roc_auc_score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "9ce43c68",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Instantiate the model\n",
    "rf = RandomForestClassifier(\n",
    "    n_estimators=500,      # you can tune this\n",
    "    max_depth=None,        # full depth; you can limit for speed/regularization\n",
    "    random_state=42,\n",
    "    n_jobs=-1              # use all cores\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "04bb89dc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>#sk-container-id-1 {\n",
       "  /* Definition of color scheme common for light and dark mode */\n",
       "  --sklearn-color-text: #000;\n",
       "  --sklearn-color-text-muted: #666;\n",
       "  --sklearn-color-line: gray;\n",
       "  /* Definition of color scheme for unfitted estimators */\n",
       "  --sklearn-color-unfitted-level-0: #fff5e6;\n",
       "  --sklearn-color-unfitted-level-1: #f6e4d2;\n",
       "  --sklearn-color-unfitted-level-2: #ffe0b3;\n",
       "  --sklearn-color-unfitted-level-3: chocolate;\n",
       "  /* Definition of color scheme for fitted estimators */\n",
       "  --sklearn-color-fitted-level-0: #f0f8ff;\n",
       "  --sklearn-color-fitted-level-1: #d4ebff;\n",
       "  --sklearn-color-fitted-level-2: #b3dbfd;\n",
       "  --sklearn-color-fitted-level-3: cornflowerblue;\n",
       "\n",
       "  /* Specific color for light theme */\n",
       "  --sklearn-color-text-on-default-background: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, black)));\n",
       "  --sklearn-color-background: var(--sg-background-color, var(--theme-background, var(--jp-layout-color0, white)));\n",
       "  --sklearn-color-border-box: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, black)));\n",
       "  --sklearn-color-icon: #696969;\n",
       "\n",
       "  @media (prefers-color-scheme: dark) {\n",
       "    /* Redefinition of color scheme for dark theme */\n",
       "    --sklearn-color-text-on-default-background: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, white)));\n",
       "    --sklearn-color-background: var(--sg-background-color, var(--theme-background, var(--jp-layout-color0, #111)));\n",
       "    --sklearn-color-border-box: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, white)));\n",
       "    --sklearn-color-icon: #878787;\n",
       "  }\n",
       "}\n",
       "\n",
       "#sk-container-id-1 {\n",
       "  color: var(--sklearn-color-text);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 pre {\n",
       "  padding: 0;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 input.sk-hidden--visually {\n",
       "  border: 0;\n",
       "  clip: rect(1px 1px 1px 1px);\n",
       "  clip: rect(1px, 1px, 1px, 1px);\n",
       "  height: 1px;\n",
       "  margin: -1px;\n",
       "  overflow: hidden;\n",
       "  padding: 0;\n",
       "  position: absolute;\n",
       "  width: 1px;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-dashed-wrapped {\n",
       "  border: 1px dashed var(--sklearn-color-line);\n",
       "  margin: 0 0.4em 0.5em 0.4em;\n",
       "  box-sizing: border-box;\n",
       "  padding-bottom: 0.4em;\n",
       "  background-color: var(--sklearn-color-background);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-container {\n",
       "  /* jupyter's `normalize.less` sets `[hidden] { display: none; }`\n",
       "     but bootstrap.min.css set `[hidden] { display: none !important; }`\n",
       "     so we also need the `!important` here to be able to override the\n",
       "     default hidden behavior on the sphinx rendered scikit-learn.org.\n",
       "     See: https://github.com/scikit-learn/scikit-learn/issues/21755 */\n",
       "  display: inline-block !important;\n",
       "  position: relative;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-text-repr-fallback {\n",
       "  display: none;\n",
       "}\n",
       "\n",
       "div.sk-parallel-item,\n",
       "div.sk-serial,\n",
       "div.sk-item {\n",
       "  /* draw centered vertical line to link estimators */\n",
       "  background-image: linear-gradient(var(--sklearn-color-text-on-default-background), var(--sklearn-color-text-on-default-background));\n",
       "  background-size: 2px 100%;\n",
       "  background-repeat: no-repeat;\n",
       "  background-position: center center;\n",
       "}\n",
       "\n",
       "/* Parallel-specific style estimator block */\n",
       "\n",
       "#sk-container-id-1 div.sk-parallel-item::after {\n",
       "  content: \"\";\n",
       "  width: 100%;\n",
       "  border-bottom: 2px solid var(--sklearn-color-text-on-default-background);\n",
       "  flex-grow: 1;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-parallel {\n",
       "  display: flex;\n",
       "  align-items: stretch;\n",
       "  justify-content: center;\n",
       "  background-color: var(--sklearn-color-background);\n",
       "  position: relative;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-parallel-item {\n",
       "  display: flex;\n",
       "  flex-direction: column;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-parallel-item:first-child::after {\n",
       "  align-self: flex-end;\n",
       "  width: 50%;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-parallel-item:last-child::after {\n",
       "  align-self: flex-start;\n",
       "  width: 50%;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-parallel-item:only-child::after {\n",
       "  width: 0;\n",
       "}\n",
       "\n",
       "/* Serial-specific style estimator block */\n",
       "\n",
       "#sk-container-id-1 div.sk-serial {\n",
       "  display: flex;\n",
       "  flex-direction: column;\n",
       "  align-items: center;\n",
       "  background-color: var(--sklearn-color-background);\n",
       "  padding-right: 1em;\n",
       "  padding-left: 1em;\n",
       "}\n",
       "\n",
       "\n",
       "/* Toggleable style: style used for estimator/Pipeline/ColumnTransformer box that is\n",
       "clickable and can be expanded/collapsed.\n",
       "- Pipeline and ColumnTransformer use this feature and define the default style\n",
       "- Estimators will overwrite some part of the style using the `sk-estimator` class\n",
       "*/\n",
       "\n",
       "/* Pipeline and ColumnTransformer style (default) */\n",
       "\n",
       "#sk-container-id-1 div.sk-toggleable {\n",
       "  /* Default theme specific background. It is overwritten whether we have a\n",
       "  specific estimator or a Pipeline/ColumnTransformer */\n",
       "  background-color: var(--sklearn-color-background);\n",
       "}\n",
       "\n",
       "/* Toggleable label */\n",
       "#sk-container-id-1 label.sk-toggleable__label {\n",
       "  cursor: pointer;\n",
       "  display: flex;\n",
       "  width: 100%;\n",
       "  margin-bottom: 0;\n",
       "  padding: 0.5em;\n",
       "  box-sizing: border-box;\n",
       "  text-align: center;\n",
       "  align-items: start;\n",
       "  justify-content: space-between;\n",
       "  gap: 0.5em;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 label.sk-toggleable__label .caption {\n",
       "  font-size: 0.6rem;\n",
       "  font-weight: lighter;\n",
       "  color: var(--sklearn-color-text-muted);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 label.sk-toggleable__label-arrow:before {\n",
       "  /* Arrow on the left of the label */\n",
       "  content: \"▸\";\n",
       "  float: left;\n",
       "  margin-right: 0.25em;\n",
       "  color: var(--sklearn-color-icon);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 label.sk-toggleable__label-arrow:hover:before {\n",
       "  color: var(--sklearn-color-text);\n",
       "}\n",
       "\n",
       "/* Toggleable content - dropdown */\n",
       "\n",
       "#sk-container-id-1 div.sk-toggleable__content {\n",
       "  max-height: 0;\n",
       "  max-width: 0;\n",
       "  overflow: hidden;\n",
       "  text-align: left;\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-0);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-toggleable__content.fitted {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-0);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-toggleable__content pre {\n",
       "  margin: 0.2em;\n",
       "  border-radius: 0.25em;\n",
       "  color: var(--sklearn-color-text);\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-0);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-toggleable__content.fitted pre {\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-0);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 input.sk-toggleable__control:checked~div.sk-toggleable__content {\n",
       "  /* Expand drop-down */\n",
       "  max-height: 200px;\n",
       "  max-width: 100%;\n",
       "  overflow: auto;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 input.sk-toggleable__control:checked~label.sk-toggleable__label-arrow:before {\n",
       "  content: \"▾\";\n",
       "}\n",
       "\n",
       "/* Pipeline/ColumnTransformer-specific style */\n",
       "\n",
       "#sk-container-id-1 div.sk-label input.sk-toggleable__control:checked~label.sk-toggleable__label {\n",
       "  color: var(--sklearn-color-text);\n",
       "  background-color: var(--sklearn-color-unfitted-level-2);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-label.fitted input.sk-toggleable__control:checked~label.sk-toggleable__label {\n",
       "  background-color: var(--sklearn-color-fitted-level-2);\n",
       "}\n",
       "\n",
       "/* Estimator-specific style */\n",
       "\n",
       "/* Colorize estimator box */\n",
       "#sk-container-id-1 div.sk-estimator input.sk-toggleable__control:checked~label.sk-toggleable__label {\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-2);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-estimator.fitted input.sk-toggleable__control:checked~label.sk-toggleable__label {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-2);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-label label.sk-toggleable__label,\n",
       "#sk-container-id-1 div.sk-label label {\n",
       "  /* The background is the default theme color */\n",
       "  color: var(--sklearn-color-text-on-default-background);\n",
       "}\n",
       "\n",
       "/* On hover, darken the color of the background */\n",
       "#sk-container-id-1 div.sk-label:hover label.sk-toggleable__label {\n",
       "  color: var(--sklearn-color-text);\n",
       "  background-color: var(--sklearn-color-unfitted-level-2);\n",
       "}\n",
       "\n",
       "/* Label box, darken color on hover, fitted */\n",
       "#sk-container-id-1 div.sk-label.fitted:hover label.sk-toggleable__label.fitted {\n",
       "  color: var(--sklearn-color-text);\n",
       "  background-color: var(--sklearn-color-fitted-level-2);\n",
       "}\n",
       "\n",
       "/* Estimator label */\n",
       "\n",
       "#sk-container-id-1 div.sk-label label {\n",
       "  font-family: monospace;\n",
       "  font-weight: bold;\n",
       "  display: inline-block;\n",
       "  line-height: 1.2em;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-label-container {\n",
       "  text-align: center;\n",
       "}\n",
       "\n",
       "/* Estimator-specific */\n",
       "#sk-container-id-1 div.sk-estimator {\n",
       "  font-family: monospace;\n",
       "  border: 1px dotted var(--sklearn-color-border-box);\n",
       "  border-radius: 0.25em;\n",
       "  box-sizing: border-box;\n",
       "  margin-bottom: 0.5em;\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-0);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-estimator.fitted {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-0);\n",
       "}\n",
       "\n",
       "/* on hover */\n",
       "#sk-container-id-1 div.sk-estimator:hover {\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-2);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-estimator.fitted:hover {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-2);\n",
       "}\n",
       "\n",
       "/* Specification for estimator info (e.g. \"i\" and \"?\") */\n",
       "\n",
       "/* Common style for \"i\" and \"?\" */\n",
       "\n",
       ".sk-estimator-doc-link,\n",
       "a:link.sk-estimator-doc-link,\n",
       "a:visited.sk-estimator-doc-link {\n",
       "  float: right;\n",
       "  font-size: smaller;\n",
       "  line-height: 1em;\n",
       "  font-family: monospace;\n",
       "  background-color: var(--sklearn-color-background);\n",
       "  border-radius: 1em;\n",
       "  height: 1em;\n",
       "  width: 1em;\n",
       "  text-decoration: none !important;\n",
       "  margin-left: 0.5em;\n",
       "  text-align: center;\n",
       "  /* unfitted */\n",
       "  border: var(--sklearn-color-unfitted-level-1) 1pt solid;\n",
       "  color: var(--sklearn-color-unfitted-level-1);\n",
       "}\n",
       "\n",
       ".sk-estimator-doc-link.fitted,\n",
       "a:link.sk-estimator-doc-link.fitted,\n",
       "a:visited.sk-estimator-doc-link.fitted {\n",
       "  /* fitted */\n",
       "  border: var(--sklearn-color-fitted-level-1) 1pt solid;\n",
       "  color: var(--sklearn-color-fitted-level-1);\n",
       "}\n",
       "\n",
       "/* On hover */\n",
       "div.sk-estimator:hover .sk-estimator-doc-link:hover,\n",
       ".sk-estimator-doc-link:hover,\n",
       "div.sk-label-container:hover .sk-estimator-doc-link:hover,\n",
       ".sk-estimator-doc-link:hover {\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-3);\n",
       "  color: var(--sklearn-color-background);\n",
       "  text-decoration: none;\n",
       "}\n",
       "\n",
       "div.sk-estimator.fitted:hover .sk-estimator-doc-link.fitted:hover,\n",
       ".sk-estimator-doc-link.fitted:hover,\n",
       "div.sk-label-container:hover .sk-estimator-doc-link.fitted:hover,\n",
       ".sk-estimator-doc-link.fitted:hover {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-3);\n",
       "  color: var(--sklearn-color-background);\n",
       "  text-decoration: none;\n",
       "}\n",
       "\n",
       "/* Span, style for the box shown on hovering the info icon */\n",
       ".sk-estimator-doc-link span {\n",
       "  display: none;\n",
       "  z-index: 9999;\n",
       "  position: relative;\n",
       "  font-weight: normal;\n",
       "  right: .2ex;\n",
       "  padding: .5ex;\n",
       "  margin: .5ex;\n",
       "  width: min-content;\n",
       "  min-width: 20ex;\n",
       "  max-width: 50ex;\n",
       "  color: var(--sklearn-color-text);\n",
       "  box-shadow: 2pt 2pt 4pt #999;\n",
       "  /* unfitted */\n",
       "  background: var(--sklearn-color-unfitted-level-0);\n",
       "  border: .5pt solid var(--sklearn-color-unfitted-level-3);\n",
       "}\n",
       "\n",
       ".sk-estimator-doc-link.fitted span {\n",
       "  /* fitted */\n",
       "  background: var(--sklearn-color-fitted-level-0);\n",
       "  border: var(--sklearn-color-fitted-level-3);\n",
       "}\n",
       "\n",
       ".sk-estimator-doc-link:hover span {\n",
       "  display: block;\n",
       "}\n",
       "\n",
       "/* \"?\"-specific style due to the `<a>` HTML tag */\n",
       "\n",
       "#sk-container-id-1 a.estimator_doc_link {\n",
       "  float: right;\n",
       "  font-size: 1rem;\n",
       "  line-height: 1em;\n",
       "  font-family: monospace;\n",
       "  background-color: var(--sklearn-color-background);\n",
       "  border-radius: 1rem;\n",
       "  height: 1rem;\n",
       "  width: 1rem;\n",
       "  text-decoration: none;\n",
       "  /* unfitted */\n",
       "  color: var(--sklearn-color-unfitted-level-1);\n",
       "  border: var(--sklearn-color-unfitted-level-1) 1pt solid;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 a.estimator_doc_link.fitted {\n",
       "  /* fitted */\n",
       "  border: var(--sklearn-color-fitted-level-1) 1pt solid;\n",
       "  color: var(--sklearn-color-fitted-level-1);\n",
       "}\n",
       "\n",
       "/* On hover */\n",
       "#sk-container-id-1 a.estimator_doc_link:hover {\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-3);\n",
       "  color: var(--sklearn-color-background);\n",
       "  text-decoration: none;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 a.estimator_doc_link.fitted:hover {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-3);\n",
       "}\n",
       "</style><div id=\"sk-container-id-1\" class=\"sk-top-container\"><div class=\"sk-text-repr-fallback\"><pre>RandomForestClassifier(n_estimators=500, n_jobs=-1, random_state=42)</pre><b>In a Jupyter environment, please rerun this cell to show the HTML representation or trust the notebook. <br />On GitHub, the HTML representation is unable to render, please try loading this page with nbviewer.org.</b></div><div class=\"sk-container\" hidden><div class=\"sk-item\"><div class=\"sk-estimator fitted sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-1\" type=\"checkbox\" checked><label for=\"sk-estimator-id-1\" class=\"sk-toggleable__label fitted sk-toggleable__label-arrow\"><div><div>RandomForestClassifier</div></div><div><a class=\"sk-estimator-doc-link fitted\" rel=\"noreferrer\" target=\"_blank\" href=\"https://scikit-learn.org/1.6/modules/generated/sklearn.ensemble.RandomForestClassifier.html\">?<span>Documentation for RandomForestClassifier</span></a><span class=\"sk-estimator-doc-link fitted\">i<span>Fitted</span></span></div></label><div class=\"sk-toggleable__content fitted\"><pre>RandomForestClassifier(n_estimators=500, n_jobs=-1, random_state=42)</pre></div> </div></div></div></div>"
      ],
      "text/plain": [
       "RandomForestClassifier(n_estimators=500, n_jobs=-1, random_state=42)"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Train\n",
    "rf.fit(X_train, y_train)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "4c8246cf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Accuracy: 82.76%\n"
     ]
    }
   ],
   "source": [
    "# Predict\n",
    "y_pred = rf.predict(X_test)\n",
    "\n",
    "n = y_test.shape[0]\n",
    "count_mispred = 0\n",
    "for i in range(n):\n",
    "    if y_test[i] != y_pred[i]:\n",
    "        count_mispred += 1\n",
    "\n",
    "# Compute the percentage of mispredictions (accuracy)\n",
    "percentage_mispred = count_mispred / n\n",
    "print(f'Accuracy: {(1-percentage_mispred)*100:.2f}%')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "ml_env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.17"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
